// @gehtsoft/LuaxStdlib-node/lib/csvParser.js
'use strict';
class StdlibCsvParser {
    valueSeparator;
    allowStrings;
    allowComments;
    commentPrefix;
    splitLine(line) {
        if (this.valueSeparator == null || this.valueSeparator.length != 1)
            throw new Error("The value separator must be exactly one character");
        if (this.allowComments && line.startsWith(this.commentPrefix))
            return null;

        let result = [];
        let inString = false;
        let value = "";
        for (let c of line) {
            if (c == this.valueSeparator && !inString) {
                result.push(value);
                value = "";
            } else if (c == '"' && this.allowStrings)
                inString = !inString;
            else {
                value += c;
            }
        }
        result.push(value);
        return result;
    }

    constructor(){
        this.allowStrings = false;
        this.allowComments = false;
        this.commentPrefix = "#";
        this.valueSeparator = ",";
    }
}

let csvParser = StdlibCsvParser;
module.exports = csvParser
