// @gehtsoft/traduAPI-node/lib/AccountCommissionBuilder.js
'use strict';
let AccountCommission = require('./AccountCommission.js');
let variantCast = require('@gehtsoft/LuaxStdlib-node').variantCast;
//@@DocBrief("The builder of the AccountCommission class")
class AccountCommissionBuilder extends AccountCommission {
    setAcctId(acctId) {
        this.acctId = acctId;
    }
    setATPId(atpId) {
        this.atpId = atpId;
    }
    setMerchantLinkId(merchantLinkId) {
        this.merchantLinkId = merchantLinkId;
    }
    setOfferId(offerId) {
        this.offerId = offerId;
    }
    setComStageString(comStageString) {
        this.comStageString = comStageString;
    }
    setComTypeString(comTypeString) {
        this.comTypeString = comTypeString;
    }
    setRebateAmount(value) {
        this.rebateAmount = variantCast.castToReal(variantCast.fromString(value));
    }
    setMinCharge(value) {
        this.minCharge = variantCast.castToReal(variantCast.fromString(value));
    }
    setLType(value) {
        this.lType = variantCast.castToInt(variantCast.fromString(value));
    }
    setChangeNumber(value) {
        this.changeNumber = variantCast.castToInt(variantCast.fromString(value));
    }
    setDiscountAmount(value) {
        this.discountAmount = variantCast.castToReal(variantCast.fromString(value));
    }
    build() {
        return this;
    }
}
module.exports = AccountCommissionBuilder
