// @gehtsoft/traduAPI-node/lib/AccountCommission.js
'use strict';
//@@DocBrief("The class provides access to offer information of an account commission")
class AccountCommission {
    acctId;
    atpId;
    merchantLinkId;
    offerId;
    comStageString;
    comTypeString;
    rebateAmount;
    minCharge;
    lType;
    changeNumber;
    discountAmount;
    constructor() {
        this.acctId = "";
        this.atpId = "";
        this.merchantLinkId = "";
        this.offerId = "";
        this.comStageString = "";
        this.comTypeString = "";
        this.rebateAmount = 0.0;
        this.minCharge = 0.0;
        this.lType = 0;
        this.changeNumber = 0;
        this.discountAmount = 0.0;
    }
    getAcctId() {
        return this.acctId;
    }
    getATPId() {
        return this.atpId;
    }
    getMerchantLinkId() {
        return this.merchantLinkId;
    }
    getOfferId() {
        return this.offerId;
    }
    getComStageString() {
        return this.comStageString;
    }
    getComTypeString() {
        return this.comTypeString;
    }
    getRebateAmount() {
        return this.rebateAmount;
    }
    getMinCharge() {
        return this.minCharge;
    }
    getLType() {
        return this.lType;
    }
    getChangeNumber() {
        return this.changeNumber;
    }
    getDiscountAmount() {
        return this.discountAmount;
    }
}
module.exports = AccountCommission
