// @gehtsoft/traduAPI-node/lib/AccountCalculatedFieldsBuilder.js
'use strict';
let AccountCalculatedFields = require('./AccountCalculatedFields.js');
class AccountCalculatedFieldsBuilder extends AccountCalculatedFields {
    setEquity(equity) {
        this.equity = equity;
    }
    setGrossPL(grossPL) {
        this.grossPL = grossPL;
    }
    setUsedMargin(usedMargin) {
        this.usedMargin = usedMargin;
    }
    setUsableMargin(usableMargin) {
        this.usableMargin = usableMargin;
    }
    setUsableMarginPercentage(usableMarginPercentage) {
        this.usableMarginPercentage = usableMarginPercentage;
    }
    setUsedMaintenanceMargin(usedMaintenanceMargin) {
        this.usedMaintenanceMargin = usedMaintenanceMargin;
    }
    setUsableMaintenanceMargin(usableMaintenanceMargin) {
        this.usableMaintenanceMargin = usableMaintenanceMargin;
    }
    setUsableMaintenanceMarginPercentage(usableMaintenanceMarginPercentage) {
        this.usableMaintenanceMarginPercentage = usableMaintenanceMarginPercentage;
    }
    setDayPL(dayPL) {
        this.dayPL = dayPL;
    }
    build() {
        return this;
    }
}
module.exports = AccountCalculatedFieldsBuilder
