// @gehtsoft/traduAPI-node/lib/AccountCalculatedFields.js
'use strict';
class AccountCalculatedFields {
    equity;
    grossPL;
    usedMargin;
    usableMargin;
    usableMarginPercentage;
    usedMaintenanceMargin;
    usableMaintenanceMargin;
    usableMaintenanceMarginPercentage;
    dayPL;
    constructor() {
        this.equity = 0.0;
        this.grossPL = 0.0;
        this.usedMargin = 0.0;
        this.usableMargin = 0.0;
        this.usableMarginPercentage = 0.0;
        this.usedMaintenanceMargin = 0.0;
        this.usableMaintenanceMargin = 0.0;
        this.usableMaintenanceMarginPercentage = 0.0;
        this.dayPL = 0.0;
    }
    getEquity() {
        return this.equity;
    }
    getGrossPL() {
        return this.grossPL;
    }
    getUsedMargin() {
        return this.usedMargin;
    }
    getUsableMargin() {
        return this.usableMargin;
    }
    getUsableMarginPercentage() {
        return this.usableMarginPercentage;
    }
    getUsedMaintenanceMargin() {
        return this.usedMaintenanceMargin;
    }
    getUsableMaintenanceMargin() {
        return this.usableMaintenanceMargin;
    }
    getUsableMaintenanceMarginPercentage() {
        return this.usableMaintenanceMarginPercentage;
    }
    getDayPL() {
        return this.dayPL;
    }
}
module.exports = AccountCalculatedFields
