// @gehtsoft/traduAPI-node/lib/Account.js
'use strict';
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
//@@DocInclude
//@@DocBrief("The class provides access to account information")
class Account {
    accountId;
    accountKind;
    accountName;
    orderAmountLimit;
    atpId;
    balance;
    baseUnitSize;
    lastMarginCallDate;
    leverageProfileId;
    m2mEquity;
    maintenanceFlag;
    maintenanceType;
    managerAccountId;
    marginCallFlag;
    refreshProfileFlags;
    nonTradeEquity;
    baseCurrency;
    baseCurrencyPrecision;
    hedgeMarginPCT;
    rolloverProfileId;
    equity;
    grossPL;
    usedMargin;
    usableMargin;
    usableMarginPercentage;
    usedMaintenanceMargin;
    usableMaintenanceMargin;
    usableMaintenanceMarginPercentage;
    dayPL;
    constructor() {
        this.accountId = "";
        this.accountKind = "";
        this.accountName = "";
        this.orderAmountLimit = 0;
        this.atpId = "";
        this.balance = 0.0;
        this.baseUnitSize = 0;
        this.lastMarginCallDate = stdlib.mkdate(1900, 1, 1);
        this.leverageProfileId = "";
        this.m2mEquity = 0.0;
        this.maintenanceFlag = false;
        this.maintenanceType = "";
        this.managerAccountId = "";
        this.marginCallFlag = "";
        this.refreshProfileFlags = "";
        this.rolloverProfileId = "";
        this.nonTradeEquity = 0.0;
        this.equity = 0.0;
        this.grossPL = 0.0;
        this.usedMargin = 0.0;
        this.usableMargin = 0.0;
        this.usableMarginPercentage = 0.0;
        this.usedMaintenanceMargin = 0.0;
        this.usableMaintenanceMargin = 0.0;
        this.usableMaintenanceMarginPercentage = 0.0;
        this.dayPL = 0.0;
        this.baseCurrency = "";
        this.baseCurrencyPrecision = 0;
        this.hedgeMarginPCT = 0.0;
    }
    getAccountId() {
        return this.accountId;
    }
    getAccountKind() {
        return this.accountKind;
    }
    getAccountName() {
        return this.accountName;
    }
    getOrderAmountLimit() {
        return this.orderAmountLimit;
    }
    getATPId() {
        return this.atpId;
    }
    getBalance() {
        return this.balance;
    }
    getBaseUnitSize() {
        return this.baseUnitSize;
    }
    getLastMarginCallDate() {
        return this.lastMarginCallDate;
    }
    getLeverageProfileId() {
        return this.leverageProfileId;
    }
    getM2MEquity() {
        return this.m2mEquity;
    }
    getMaintenanceFlag() {
        return this.maintenanceFlag;
    }
    getMaintenanceType() {
        return this.maintenanceType;
    }
    getManagerAccountId() {
        return this.managerAccountId;
    }
    getMarginCallFlag() {
        return this.marginCallFlag;
    }
    getRefreshProfileFlags() {
        return this.refreshProfileFlags;
    }
    getNonTradeEquity() {
        return this.nonTradeEquity;
    }
    getBaseCurrency() {
        return this.baseCurrency;
    }
    getBaseCurrencyPrecision() {
        return this.baseCurrencyPrecision;
    }
    getHedgeMarginPCT() {
        return this.hedgeMarginPCT;
    }
    getRolloverProfileId() {
        return this.rolloverProfileId;
    }
    getEquity() {
        return this.equity;
    }
    getGrossPL() {
        return this.grossPL;
    }
    getUsedMargin() {
        return this.usedMargin;
    }
    getUsableMargin() {
        return this.usableMargin;
    }
    getUsableMarginPercentage() {
        return this.usableMarginPercentage;
    }
    getUsedMaintenanceMargin() {
        return this.usedMaintenanceMargin;
    }
    getUsableMaintenanceMargin() {
        return this.usableMaintenanceMargin;
    }
    getUsableMaintenanceMarginPercentage() {
        return this.usableMaintenanceMarginPercentage;
    }
    getDayPL() {
        return this.dayPL;
    }
}
module.exports = Account
