// @gehtsoft/traduAPI-node/lib/APdasGetDataRequestMessage.js
'use strict';
let APdasRequestMessage = require('./APdasRequestMessage.js');
let PdasMessageFieldFactory = require('./PdasMessageFieldFactory.js');
let PdasMessageSerializer = require('./PdasMessageSerializer.js');
let PdasMessageFieldTag = require('./PdasMessageFieldTag.js');
let PdasMessageValue = require('./PdasMessageValue.js');
let PdasConstants = require('./PdasConstants.js');
let PdasTransportMessage = require('./PdasTransportMessage.js');
class APdasGetDataRequestMessage extends APdasRequestMessage {
    requestNumberGenerator;
    build(tradingSession, requestNumberGenerator, transactTime) {
        let pdasMessageSerializer, factory, pdasMessage, requestId, pdasMessageList, group, additionalGroups, i, result, terminalUrl;
        this.requestNumberGenerator = requestNumberGenerator;
        factory = new PdasMessageFieldFactory();
        pdasMessageSerializer = new PdasMessageSerializer();
        terminalUrl = tradingSession.getTradingTerminal().getUrls()[0];
        requestId = tradingSession.getSessionId() + "-" + (this.getRequestNumber()).toString();
        pdasMessage = APdasRequestMessage.createMessage(factory, tradingSession, transactTime, requestId);
        pdasMessageList = factory.createList(PdasMessageFieldTag.FXCM_NO_PARAM);
        group = factory.createGroup();
        group.addChild(factory.createStringField(PdasMessageFieldTag.FXCM_PARAM_NAME, PdasMessageValue.CLIENT_TRADER_ID));
        group.addChild(factory.createStringField(PdasMessageFieldTag.FXCM_PARAM_VALUE, tradingSession.getTraderId()));
        pdasMessageList.addChild(group);
        additionalGroups = this.getAdditionalGroups(factory);
        if (additionalGroups != null) {
            for (i = 0; i <= additionalGroups.length - 1; i += 1) {
                pdasMessageList.addChild(additionalGroups[i]);
            }
        }
        group = factory.createGroup();
        group.addChild(factory.createStringField(PdasMessageFieldTag.FXCM_PARAM_NAME, PdasMessageValue.CLIENT_QID));
        group.addChild(factory.createStringField(PdasMessageFieldTag.FXCM_PARAM_VALUE, requestId));
        pdasMessageList.addChild(group);
        pdasMessage.addChild(pdasMessageList);
        pdasMessage.addChild(factory.createStringField(PdasMessageFieldTag.FXCM_COMMAND_ID, this.getCommand()));
        pdasMessage.addChild(factory.createStringField(PdasConstants.SID, tradingSession.getSessionId()));
        result = PdasTransportMessage.create(this.getTransportMessageType(), this.createRequestUrl(terminalUrl), pdasMessageSerializer.serialize(pdasMessage));
        return result;
    }
    getRequestNumber() {
        return this.requestNumberGenerator.getNextRequestNumber();
    }
    getAdditionalGroups(factory) {
        return null;
    }
    getCommand() {
        return null;
    }
    getTransportMessageType() {
        return null;
    }
}
module.exports = APdasGetDataRequestMessage
