// @gehtsoft/traduAPI-node/lib/ChangeOrderCommandBuilder.js
'use strict';
let ChangeOrderCommand = require('./ChangeOrderCommand.js');
class ChangeOrderCommandBuilder extends ChangeOrderCommand {
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setTradingSession(tradingSession) {
        this.tradingSession = tradingSession;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setAccountId(accountId) {
        this.accountId = accountId;
    }
    setOrderId(orderId) {
        this.orderId = orderId;
    }
    setOrderType(orderType) {
        this.orderType = orderType;
    }
    setChangeOrderRequest(changeOrderRequest) {
        this.changeOrderRequest = changeOrderRequest;
    }
    setPointSize(pointSize) {
        this.pointSize = pointSize;
    }
    build() {
        return this;
    }
}
module.exports = ChangeOrderCommandBuilder
