// @gehtsoft/traduAPI-node/lib/ChangeOrderRequest.js
'use strict';
let TrailingStopType = require('./TrailingStopType.js');
//@@DocInclude
//@@DocBrief("The class provides order change data")
class ChangeOrderRequest {
    trailingType;
    trailingStep;
    amount;
    orderId;
    rate;
    pips;
    rateRange;
    constructor() {
        this.amount = 0;
        this.rate = 0.0;
        this.orderId = "";
        this.rateRange = 0.0;
        this.pips = 0.0;
        this.trailingType = TrailingStopType.None;
        this.trailingStep = 0;
    }
    isEnabledTrailing() {
        return this.trailingType != TrailingStopType.None;
    }
    getTrailingType() {
        return this.trailingType;
    }
    getTrailingStep() {
        return this.trailingStep;
    }
    getAmount() {
        return this.amount;
    }
    getOrderId() {
        return this.orderId;
    }
    getRate() {
        return this.rate;
    }
    getPips() {
        return this.pips;
    }
    getRateRange() {
        return this.rateRange;
    }
}
module.exports = ChangeOrderRequest
