// @gehtsoft/traduAPI-node/lib/ClientMessagesStorage.js
'use strict';
let string_map = require('@gehtsoft/LuaxStdlib-node').string_map;
let list = require('@gehtsoft/LuaxStdlib-node').list;
class ClientMessagesStorage {
    storage;
    constructor() {
        this.storage = new string_map();
    }
    getAllMessages() {
        let length, i, keys, messages, message, result;
        keys = this.storage.keys();
        length = keys.length;
        messages = new list();
        for (i = 0; i <= length - 1; i += 1) {
            message = this.getMessageById(keys[i]);
            messages.add(message);
        }
        result = new Array(messages.length());
        for (i = 0; i <= messages.length() - 1; i += 1) {
            result[i] = messages.get(i);
        }
        return result;
    }
    addMessage(message) {
        this.storage.set(message.getMessageId(), message);
    }
    getMessageById(id) {
        if (!this.storage.contains(id)) {
            return null;
        }
        return this.storage.get(id);
    }
    clear() {
        this.storage.clear();
    }
}
module.exports = ClientMessagesStorage
