// @gehtsoft/traduAPI-node/lib/CloseMarketOrderRequestBuilder.js
'use strict';
let CloseMarketOrderRequest = require('./CloseMarketOrderRequest.js');
//@@DocInclude
//@@DocBrief("The builder of the CloseMarketOrderRequest class")
class CloseMarketOrderRequestBuilder extends CloseMarketOrderRequest {
    constructor() {
        super();
        this.customId = "";
        this.tradeId = "";
    }
    setTradeId(tradeId) {
        this.tradeId = tradeId;
        return this;
    }
    setAmount(amount) {
        this.amount = amount;
        return this;
    }
    setTimeInForce(timeInForce) {
        this.timeInForce = timeInForce;
        return this;
    }
    setExpirationDate(expirationDate) {
        this.expirationDate = expirationDate;
        return this;
    }
    setCustomId(customId) {
        this.customId = customId;
        return this;
    }
    setRateRange(rateRange) {
        this.rateRange = rateRange;
        this.rateRangeFilled = true;
        return this;
    }
    build() {
        return this;
    }
}
module.exports = CloseMarketOrderRequestBuilder
