// @gehtsoft/traduAPI-node/lib/CloseAllPositionsRequestBuilder.js
'use strict';
let CloseAllPositionsRequest = require('./CloseAllPositionsRequest.js');
//@@DocInclude
//@@DocBrief("The builder for the CloseAllPositionsRequest")
class CloseAllPositionsRequestBuilder extends CloseAllPositionsRequest {
    setTimeInForce(timeInForce) {
        this.timeInForce = timeInForce;
        return this;
    }
    setExpirationDate(expirationDate) {
        this.expirationDate = expirationDate;
        return this;
    }
    setAcctId(acctId) {
        this.acctId = acctId;
        return this;
    }
    setCustomId(customId) {
        this.customId = customId;
        return this;
    }
    setRateRange(rateRange) {
        this.rateRange = rateRange;
        this.rateRangeFilled = true;
        return this;
    }
    setSide(side) {
        this.side = side;
        this.sideFilled = true;
        return this;
    }
    setSymbol(symbol) {
        this.symbol = symbol;
        this.symbolFilled = true;
        return this;
    }
    build() {
        return this;
    }
}
module.exports = CloseAllPositionsRequestBuilder
