// @gehtsoft/traduAPI-node/lib/CloseAllPositionsRequest.js
'use strict';
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
//@@DocInclude
//@@DocBrief("The class provides information for a close all positions market order request")
class CloseAllPositionsRequest {
    timeInForce;
    expirationDate;
    acctId;
    customId;
    rateRange;
    rateRangeFilled;
    side;
    sideFilled;
    symbol;
    symbolFilled;
    constructor() {
        this.rateRange = 0.0;
        this.acctId = "";
        this.timeInForce = "GTC";
        this.expirationDate = stdlib.mkdate(1900, 1, 1);
        this.customId = "";
        this.rateRangeFilled = false;
        this.side = "";
        this.symbol = "";
        this.sideFilled = false;
        this.symbolFilled = false;
    }
    getTimeInForce() {
        return this.timeInForce;
    }
    getExpirationDate() {
        return this.expirationDate;
    }
    getAcctId() {
        return this.acctId;
    }
    getCustomId() {
        return this.customId;
    }
    getRateRange() {
        return this.rateRange;
    }
    isRateRangeFilled() {
        return this.rateRangeFilled;
    }
    getSide() {
        return this.side;
    }
    isSideFilled() {
        return this.sideFilled;
    }
    getSymbol() {
        return this.symbol;
    }
    isSymbolFilled() {
        return this.symbolFilled;
    }
}
module.exports = CloseAllPositionsRequest
