// @gehtsoft/traduAPI-node/lib/ClosedPosition.js
'use strict';
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
//@@DocInclude
//@@DocBrief("The class provides access to information about a closed position")
class ClosedPosition {
    tradeID;
    accountId;
    accountName;
    accountKind;
    offerId;
    amount;
    buySell;
    openRate;
    openTime;
    openQuoteId;
    openOrderId;
    openOrderReqId;
    openOrderRequestTXT;
    closeRate;
    closeTime;
    closeQuoteId;
    closeOrderId;
    closeOrderReqId;
    closeOrderRequestTXT;
    parties;
    commission;
    rolloverInterest;
    tradeIdOrigin;
    tradeIdRemain;
    valueDate;
    dividends;
    pl;
    plPips;
    netPL;
    grossPL;
    constructor() {
        this.tradeID = null;
        this.accountId = null;
        this.accountName = null;
        this.accountKind = null;
        this.offerId = null;
        this.amount = 0;
        this.buySell = null;
        this.openRate = 0.0;
        this.openTime = stdlib.nowlocal();
        this.openQuoteId = null;
        this.openOrderId = null;
        this.openOrderReqId = null;
        this.openOrderRequestTXT = null;
        this.closeRate = 0.0;
        this.closeTime = stdlib.nowlocal();
        this.closeQuoteId = null;
        this.closeOrderId = null;
        this.closeOrderReqId = null;
        this.closeOrderRequestTXT = null;
        this.commission = 0.0;
        this.rolloverInterest = 0.0;
        this.tradeIdOrigin = null;
        this.tradeIdRemain = null;
        this.valueDate = null;
        this.parties = null;
        this.pl = 0;
        this.plPips = 0;
        this.grossPL = 0;
        this.netPL = 0;
    }
    getTradeID() {
        return this.tradeID;
    }
    getAccountId() {
        return this.accountId;
    }
    getAccountName() {
        return this.accountName;
    }
    getAccountKind() {
        return this.accountKind;
    }
    getOfferId() {
        return this.offerId;
    }
    getAmount() {
        return this.amount;
    }
    getBuySell() {
        return this.buySell;
    }
    getOpenRate() {
        return this.openRate;
    }
    getOpenTime() {
        return this.openTime;
    }
    getOpenQuoteId() {
        return this.openQuoteId;
    }
    getOpenOrderId() {
        return this.openOrderId;
    }
    getOpenOrderReqId() {
        return this.openOrderReqId;
    }
    getOpenOrderRequestTXT() {
        return this.openOrderRequestTXT;
    }
    getCloseRate() {
        return this.closeRate;
    }
    getCloseTime() {
        return this.closeTime;
    }
    getCloseQuoteId() {
        return this.closeQuoteId;
    }
    getCloseOrderId() {
        return this.closeOrderId;
    }
    getCloseOrderReqId() {
        return this.closeOrderReqId;
    }
    getCloseOrderRequestTXT() {
        return this.closeOrderRequestTXT;
    }
    getCloseOrderParties() {
        return this.parties;
    }
    getCommission() {
        return this.commission;
    }
    getRolloverInterest() {
        return this.rolloverInterest;
    }
    getTradeIdOrigin() {
        return this.tradeIdOrigin;
    }
    getTradeIdRemain() {
        return this.tradeIdRemain;
    }
    getValueDate() {
        return this.valueDate;
    }
    getDividends() {
        return this.dividends;
    }
    getPL() {
        return this.pl;
    }
    getPLPips() {
        return this.plPips;
    }
    getNetPL() {
        return this.netPL;
    }
    getGrossPL() {
        return this.grossPL;
    }
}
module.exports = ClosedPosition
