// @gehtsoft/traduAPI-node/lib/ClosedPositionUpdate.js
'use strict';
//@@DocBrief("The class provides access to information about a closed position update")
class ClosedPositionUpdate {
    ClosedPosition;
    tradeIdChanged;
    accountIdChanged;
    accountNameChanged;
    accountKindChanged;
    offerIdChanged;
    amountChanged;
    buySellChanged;
    openRateChanged;
    openTimeChanged;
    openQuoteIdChanged;
    openOrderIdChanged;
    openOrderReqIdChanged;
    openOrderRequestTXTChanged;
    closeRateChanged;
    closeTimeChanged;
    closeQuoteIdChanged;
    closeOrderIdChanged;
    closeOrderReqIdChanged;
    closeOrderRequestTXTChanged;
    partiesChanged;
    grossPLChanged;
    commissionChanged;
    rolloverInterestChanged;
    tradeIdOriginChanged;
    tradeIdRemainChanged;
    valueDateChanged;
    dividendsChanged;
    getClosedPosition() {
        return this.ClosedPosition;
    }
    isTradeIdChanged() {
        return this.tradeIdChanged;
    }
    isAccountIdChanged() {
        return this.accountIdChanged;
    }
    isAccountNameChanged() {
        return this.accountNameChanged;
    }
    isAccountKindChanged() {
        return this.accountKindChanged;
    }
    isOfferIdChanged() {
        return this.offerIdChanged;
    }
    isAmountChanged() {
        return this.amountChanged;
    }
    isBuySellChanged() {
        return this.buySellChanged;
    }
    isOpenRateChanged() {
        return this.openRateChanged;
    }
    isOpenTimeChanged() {
        return this.openTimeChanged;
    }
    isOpenQuoteIdChanged() {
        return this.openQuoteIdChanged;
    }
    isOpenOrderIdChanged() {
        return this.openOrderIdChanged;
    }
    isOpenOrderReqIdChanged() {
        return this.openOrderReqIdChanged;
    }
    isOpenOrderRequestTXTChanged() {
        return this.openOrderRequestTXTChanged;
    }
    isCloseRateChanged() {
        return this.closeRateChanged;
    }
    isCloseTimeChanged() {
        return this.closeTimeChanged;
    }
    isCloseQuoteIdChanged() {
        return this.closeQuoteIdChanged;
    }
    isCloseOrderIdChanged() {
        return this.closeOrderIdChanged;
    }
    isCloseOrderReqIdChanged() {
        return this.closeOrderReqIdChanged;
    }
    isCloseOrderRequestTXTChanged() {
        return this.closeOrderRequestTXTChanged;
    }
    isCloseOrderPartiesChanged() {
        return this.partiesChanged;
    }
    isGrossPLChanged() {
        return this.grossPLChanged;
    }
    isCommissionChanged() {
        return this.commissionChanged;
    }
    isRolloverInterestChanged() {
        return this.rolloverInterestChanged;
    }
    isTradeIdOriginChanged() {
        return this.tradeIdOriginChanged;
    }
    isTradeIdRemainChanged() {
        return this.tradeIdRemainChanged;
    }
    isValueDateChanged() {
        return this.valueDateChanged;
    }
    isDividendsChanged() {
        return this.dividendsChanged;
    }
}
module.exports = ClosedPositionUpdate
