// @gehtsoft/traduAPI-node/lib/ClosedPositionBuilder.js
'use strict';
let ClosedPosition = require('./ClosedPosition.js');
//@@DocBrief("The builder of the ClosedPosition class")
class ClosedPositionBuilder extends ClosedPosition {
    setTradeID(value) {
        this.tradeID = value;
    }
    setAccountId(value) {
        this.accountId = value;
    }
    setAccountName(value) {
        this.accountName = value;
    }
    setAccountKind(value) {
        this.accountKind = value;
    }
    setOfferId(value) {
        this.offerId = value;
    }
    setAmount(value) {
        this.amount = value;
    }
    setBuySell(value) {
        this.buySell = value;
    }
    setOpenRate(value) {
        this.openRate = value;
    }
    setOpenTime(value) {
        this.openTime = value;
    }
    setOpenQuoteId(value) {
        this.openQuoteId = value;
    }
    setOpenOrderId(value) {
        this.openOrderId = value;
    }
    setOpenOrderReqId(value) {
        this.openOrderReqId = value;
    }
    setOpenOrderRequestTXT(value) {
        this.openOrderRequestTXT = value;
    }
    setCloseRate(value) {
        this.closeRate = value;
    }
    setCloseTime(value) {
        this.closeTime = value;
    }
    setCloseQuoteId(value) {
        this.closeQuoteId = value;
    }
    setCloseOrderId(value) {
        this.closeOrderId = value;
    }
    setCloseOrderReqId(value) {
        this.closeOrderReqId = value;
    }
    setCloseOrderRequestTXT(value) {
        this.closeOrderRequestTXT = value;
    }
    setCloseOrderParties(value) {
        this.parties = value;
    }
    setCommission(value) {
        this.commission = value;
    }
    setRolloverInterest(value) {
        this.rolloverInterest = value;
    }
    setTradeIdOrigin(value) {
        this.tradeIdOrigin = value;
    }
    setTradeIdRemain(value) {
        this.tradeIdRemain = value;
    }
    setValueDate(value) {
        this.valueDate = value;
    }
    setGrossPL(value) {
        this.grossPL = value;
    }
    setDividends(value) {
        this.dividends = value;
    }
    setNetPL(value) {
        this.netPL = value;
    }
    setPL(value) {
        this.pl = value;
    }
    setPLPips(value) {
        this.plPips = value;
    }
    build() {
        return this;
    }
}
module.exports = ClosedPositionBuilder
