// @gehtsoft/traduAPI-node/lib/CollectionUtils.js
'use strict';
let StringValueObject = require('./StringValueObject.js');
class CollectionUtils {
    static addStringToList(collection, value) {
        let i;
        for (i = 0; i <= collection.length() - 1; i += 1) {
            if (collection.get(i).get() == value) {
                return;
            }
        }
        collection.add(StringValueObject.create(value));
    }
    static removeStringFromList(collection, value) {
        let i;
        for (i = 0; i <= collection.length() - 1; i += 1) {
            if (collection.get(i).get() == value) {
                collection.remove(i);
                return;
            }
        }
    }
    static containsStringInList(collection, value) {
        let i;
        for (i = 0; i <= collection.length() - 1; i += 1) {
            if (collection.get(i).get() == value) {
                return true;
            }
        }
        return false;
    }
    static stringListToArray(collection) {
        let i, result;
        result = new Array(collection.length());
        for (i = 0; i <= collection.length() - 1; i += 1) {
            result[i] = collection.get(i).get();
        }
        return result;
    }
    static clearQueue(q) {
        while(q.length() > 0) {
            q.dequeue();
        }
    }
}
module.exports = CollectionUtils
