// @gehtsoft/traduAPI-node/lib/CommandEnvironment.js
'use strict';
//@@DocBrief("The class of parameters to generate commands")
class CommandEnvironment {
    messageRouter;
    messageExecutor;
    requestNumberGenerator;
    messageFactory;
    tradingTerminal;
    connectionParameters;
    configFlags;
    constructor() {
        this.messageRouter = null;
        this.messageExecutor = null;
        this.requestNumberGenerator = null;
        this.messageFactory = null;
        this.tradingTerminal = null;
        this.connectionParameters = null;
    }
    getMessageRouter() {
        return this.messageRouter;
    }
    getMessageExecutor() {
        return this.messageExecutor;
    }
    getRequestNumberGenerator() {
        return this.requestNumberGenerator;
    }
    getMessageFactory() {
        return this.messageFactory;
    }
    getTradingTerminal() {
        return this.tradingTerminal;
    }
    getConnectionParameters() {
        return this.connectionParameters;
    }
    getConfigFlags() {
        return this.configFlags;
    }
}
module.exports = CommandEnvironment
