// @gehtsoft/traduAPI-node/lib/ConnectionParameters.js
'use strict';
let ResilienceBuilder = require('./ResilienceBuilder.js');
//@@DocInclude
//@@DocBrief("An instance of this class stores connection parameters")
class ConnectionParameters {
    reconnectionsNumber;
    reconnectionDelayInSeconds;
    loginTimeoutInSeconds;
    applicationName;
    agent;
    resilience;
    client;
    constructor() {
        let resilienceBuilder;
        resilienceBuilder = new ResilienceBuilder();
        resilienceBuilder.setAttempts(10);
        resilienceBuilder.setPeriodInMilliseconds(500);
        this.reconnectionsNumber = 10;
        this.reconnectionDelayInSeconds = 3;
        this.loginTimeoutInSeconds = 120;
        this.applicationName = "";
        this.agent = "traduAPI";
        this.resilience = resilienceBuilder.build();
    }
    getLoginTimeoutInSeconds() {
        return this.loginTimeoutInSeconds;
    }
    getApplicationName() {
        return this.applicationName;
    }
    getAgent() {
        return this.agent;
    }
    getReconnectionsNumber() {
        return this.reconnectionsNumber;
    }
    getReconnectionDelayInSeconds() {
        return this.reconnectionDelayInSeconds;
    }
    getResilience() {
        return this.resilience;
    }
}
module.exports = ConnectionParameters
