// @gehtsoft/traduAPI-node/lib/ConnectionStatus.js
'use strict';
//@@DocBrief("The class for checking the current connection status of the session")
class ConnectionStatus {
    static get DISCONNECTED() { return 0; }
    static get CONNECTING() { return 1; }
    static get CONNECTED() { return 2; }
    static get RECONNECTING() { return 3; }
    status;
    error;
    constructor() {
        this.status = 0;
        this.error = null;
    }
    setStatus(status) {
        this.status = status;
        this.error = null;
    }
    setStatusWithError(status, error) {
        this.status = status;
        this.error = error;
    }
    hasError() {
        return this.error != null;
    }
    getError() {
        return this.error;
    }
    isConnected() {
        return this.status == ConnectionStatus.CONNECTED;
    }
    isDisconnected() {
        return this.status == ConnectionStatus.DISCONNECTED;
    }
    isConnecting() {
        return this.status == ConnectionStatus.CONNECTING;
    }
    isReconnecting() {
        return this.status == ConnectionStatus.RECONNECTING;
    }
}
module.exports = ConnectionStatus
