// @gehtsoft/traduAPI-node/lib/ControllersBreaker.js
'use strict';
let list = require('@gehtsoft/LuaxStdlib-node').list;
let LogManager = require('./LogManager.js');
class ControllersBreaker {
    controllersStorage;
    logger;
    constructor() {
        this.controllersStorage = new list();
        this.logger = LogManager.getLogger();
    }
    add(controller) {
        let i, storageItem;
        for (i = 0; i <= this.controllersStorage.length() - 1; i += 1) {
            storageItem = this.controllersStorage.get(i);
            if (controller == storageItem) {
                return;
            }
        }
        this.controllersStorage.add(controller);
    }
    remove(controller) {
        let i, storageItem;
        for (i = 0; i <= this.controllersStorage.length() - 1; i += 1) {
            storageItem = this.controllersStorage.get(i);
            if (controller == storageItem) {
                this.controllersStorage.remove(i);
                break;
            }
        }
    }
    stopAll() {
        let i, stoppableController;
        this.logger.debug("ControllersBreaker. Stop all");
        for (i = 0; i <= this.controllersStorage.length() - 1; i += 1) {
            stoppableController = this.controllersStorage.get(i);
            stoppableController.stop();
        }
    }
}
module.exports = ControllersBreaker
