// @gehtsoft/traduAPI-node/lib/DefaultLogger.js
'use strict';
let logger = require('@gehtsoft/LuaxStdlib-node').logger;
class DefaultLogger {
    level;
    constructor() {
        this.level = -1;
    }
    setLevel(level) {
        if ("debug" == level) {
            this.level = 0;
        } else if ("info" == level) {
            this.level = 1;
        } else if ("warning" == level) {
            this.level = 2;
        } else if ("error" == level) {
            this.level = 3;
        } else if ("fatal" == level) {
            this.level = 4;
        }
    }
    debug(message) {
        if (0 >= this.level) {
            logger.debug(message);
        }
    }
    info(message) {
        if (1 >= this.level) {
            logger.info(message);
        }
    }
    warning(message) {
        if (2 >= this.level) {
            logger.warning(message);
        }
    }
    error(message) {
        if (3 >= this.level) {
            logger.error(message);
        }
    }
    fatal(message) {
        if (4 >= this.level) {
            logger.fatal(message);
        }
    }
}
module.exports = DefaultLogger
