// @gehtsoft/traduAPI-node/lib/DXFeedUnsubscribeForTimeSeriesMessage.js
'use strict';
class DXFeedUnsubscribeForTimeSeriesMessage {
    id;
    message;
    constructor() {
        this.message = "";
        this.id = -1;
    }
    getId() {
        return this.id;
    }
    getMessage() {
        return this.message;
    }
    static create(clientId, requests, requestNumberGenerator) {
        let result, symbolsStr, timeframe, symbol, i;
        result = new DXFeedUnsubscribeForTimeSeriesMessage();
        result.id = requestNumberGenerator.getNextRequestNumber();
        symbolsStr = "[";
        for (i = 0; i <= requests.length - 1; i += 1) {
            timeframe = requests[i].getTimeframe().toString();
            symbol = requests[i].getSymbol();
            symbolsStr = symbolsStr + "\"" + symbol + "{=" + timeframe + "}\"";
            if (i != requests.length - 1) {
                symbolsStr = symbolsStr + ",";
            }
        }
        symbolsStr = symbolsStr + "]";
        result.message = "[{" + "\"id\":\"" + (result.id).toString() + "\"," + "\"channel\":\"/service/sub\"," + "\"data\":{\"removeTimeSeries\":{\"Candle\":" + symbolsStr + "}}," + "\"clientId\":\"" + clientId + "\"" + "}]";
        return result;
    }
}
module.exports = DXFeedUnsubscribeForTimeSeriesMessage
