// @gehtsoft/traduAPI-node/lib/DXFeedMessageFactory.js
'use strict';
let RequestNumberGenerator = require('./RequestNumberGenerator.js');
let DXFeedAuthorizeMessage = require('./DXFeedAuthorizeMessage.js');
let DXFeedHeartbeatMessage = require('./DXFeedHeartbeatMessage.js');
let DXFeedSubscribeForQuotesMessage = require('./DXFeedSubscribeForQuotesMessage.js');
let DXFeedUnsubscribeForQuotesMessage = require('./DXFeedUnsubscribeForQuotesMessage.js');
let DXFeedSubscribeForTimeSeriesMessage = require('./DXFeedSubscribeForTimeSeriesMessage.js');
let DXFeedUnsubscribeForTimeSeriesMessage = require('./DXFeedUnsubscribeForTimeSeriesMessage.js');
class DXFeedMessageFactory {
    requestNumberGenerator;
    constructor() {
        this.requestNumberGenerator = new RequestNumberGenerator();
    }
    createAuthorizeMessage(token) {
        return DXFeedAuthorizeMessage.create(token, this.requestNumberGenerator);
    }
    createHeartbeatMessage(clientId) {
        return DXFeedHeartbeatMessage.create(clientId, this.requestNumberGenerator);
    }
    createSubscribeForQuotesMessage(clientId, symbols) {
        return DXFeedSubscribeForQuotesMessage.create(clientId, symbols, this.requestNumberGenerator);
    }
    createUnsubscribeFromQuotesMessage(clientId, symbols) {
        return DXFeedUnsubscribeForQuotesMessage.create(clientId, symbols, this.requestNumberGenerator);
    }
    createSubscribeForTimeSeriesMessage(clientId, requests) {
        return DXFeedSubscribeForTimeSeriesMessage.create(clientId, requests, this.requestNumberGenerator);
    }
    createUnsubscribeForTimeSeriesMessage(clientId, requests) {
        return DXFeedUnsubscribeForTimeSeriesMessage.create(clientId, requests, this.requestNumberGenerator);
    }
}
module.exports = DXFeedMessageFactory
