// @gehtsoft/traduAPI-node/lib/DxFeedReceiveQuotesPublisher.js
'use strict';
let CommonPublisher = require('./CommonPublisher.js');
class DxFeedReceiveQuotesPublisher {
    publisher;
    constructor() {
        this.publisher = new CommonPublisher();
    }
    subscribe(callback) {
        this.publisher.subscribe(callback);
    }
    unsubscribe(callback) {
        this.publisher.unsubscribe(callback);
    }
    notifyReceivedQuotes(dXFeedQuotes) {
        this.publisher.notifyAction(DxFeedReceiveQuotesPublisher.ReceivedQuotesAction.create(dXFeedQuotes));
    }
}
module.exports = DxFeedReceiveQuotesPublisher
DxFeedReceiveQuotesPublisher.ReceivedQuotesAction = class {
    listener;
    dXFeedQuotes;
    _owner_;
    constructor(_owner_) {
        this._owner_ = _owner_;
    }
    static create(dXFeedQuotes) {
        let result;
        result = new DxFeedReceiveQuotesPublisher.ReceivedQuotesAction(this);
        result.dXFeedQuotes = dXFeedQuotes;
        return result;
    }
    setListener(listener) {
        this.listener = listener;
    }
    execute() {
        this.listener.onQuotesReceived(this.dXFeedQuotes);
    }
}
