// @gehtsoft/traduAPI-node/lib/DxFeedHandshakeData.js
'use strict';
let ADXFeedMessage = require('./ADXFeedMessage.js');
class DxFeedHandshakeData extends ADXFeedMessage {
    static get CHANNEL() { return "/meta/handshake"; }
    error;
    clientId;
    heartbeatInterval;
    sessionTimeout;
    successful;
    constructor() {
        super();
        this.channel = DxFeedHandshakeData.CHANNEL;
    }
    getError() {
        return this.error;
    }
    getClientId() {
        return this.clientId;
    }
    getSuccessful() {
        return this.successful;
    }
    getSessionTimeout() {
        return this.sessionTimeout;
    }
    getHeartbeatInterval() {
        return this.heartbeatInterval;
    }
    static create(successful, clientId, heartbeatInterval, sessionTimeout, error) {
        let dxFeedHandshakeData;
        dxFeedHandshakeData = new DxFeedHandshakeData();
        dxFeedHandshakeData.clientId = clientId;
        dxFeedHandshakeData.heartbeatInterval = heartbeatInterval;
        dxFeedHandshakeData.sessionTimeout = sessionTimeout;
        dxFeedHandshakeData.successful = successful;
        dxFeedHandshakeData.error = error;
        return dxFeedHandshakeData;
    }
}
module.exports = DxFeedHandshakeData
