// @gehtsoft/traduAPI-node/lib/DXFeedSubscribeForTimeSeriesMessage.js
'use strict';
let DateTimeParser = require('./DateTimeParser.js');
let TimeframeUnit = require('./TimeframeUnit.js');
class DXFeedSubscribeForTimeSeriesMessage {
    id;
    message;
    constructor() {
        this.message = "";
        this.id = -1;
    }
    getId() {
        return this.id;
    }
    getMessage() {
        return this.message;
    }
    static create(clientId, requests, requestNumberGenerator) {
        let result, symbolsStr, timeframe, time, symbol, i;
        result = new DXFeedSubscribeForTimeSeriesMessage();
        result.id = requestNumberGenerator.getNextRequestNumber();
        symbolsStr = "[";
        for (i = 0; i <= requests.length - 1; i += 1) {
            timeframe = DXFeedSubscribeForTimeSeriesMessage.getTimeframeString(requests[i].getTimeframe());
            time = DateTimeParser.toUnixTimestampString(requests[i].getFrom());
            symbol = requests[i].getSymbol();
            symbolsStr = symbolsStr + "{\"eventSymbol\":\"" + symbol + "{=" + timeframe + "}\",\"fromTime\":" + time + "}";
            if (i != requests.length - 1) {
                symbolsStr = symbolsStr + ",";
            }
        }
        symbolsStr = symbolsStr + "]";
        result.message = "[{" + "\"id\":\"" + (result.id).toString() + "\"," + "\"channel\":\"/service/sub\"," + "\"data\":{\"addTimeSeries\":{\"Candle\":" + symbolsStr + "}}," + "\"clientId\":\"" + clientId + "\"" + "}]";
        return result;
    }
    static getTimeframeString(timeframe) {
        if (timeframe.getUnit() == TimeframeUnit.Month) {
            return (timeframe.getSize()).toString() + "mo";
        }
        else {
            return timeframe.toString();
        }
    }
}
module.exports = DXFeedSubscribeForTimeSeriesMessage
