// @gehtsoft/traduAPI-node/lib/DXFeedUnsubscribeForQuotesMessage.js
'use strict';
class DXFeedUnsubscribeForQuotesMessage {
    id;
    message;
    constructor() {
        this.message = "";
        this.id = -1;
    }
    getId() {
        return this.id;
    }
    getMessage() {
        return this.message;
    }
    static create(clientId, symbols, requestNumberGenerator) {
        let result, symbolsStr, i;
        result = new DXFeedUnsubscribeForQuotesMessage();
        result.id = requestNumberGenerator.getNextRequestNumber();
        symbolsStr = "[";
        for (i = 0; i <= symbols.length - 1; i += 1) {
            symbolsStr = symbolsStr + "\"" + symbols[i] + "\"";
            if (i != symbols.length - 1) {
                symbolsStr = symbolsStr + ",";
            }
        }
        symbolsStr = symbolsStr + "]";
        result.message = "[{" + "\"id\":\"" + (result.id).toString() + "\"," + "\"channel\":\"/service/sub\"," + "\"data\":{\"remove\":{\"Quote\":" + symbolsStr + "}}," + "\"clientId\":\"" + clientId + "\"" + "}]";
        return result;
    }
}
module.exports = DXFeedUnsubscribeForQuotesMessage
