// @gehtsoft/traduAPI-node/lib/DXFeedTimeSeriaBuilder.js
'use strict';
let DXFeedTimeSeria = require('./DXFeedTimeSeria.js');
class DXFeedTimeSeriaBuilder extends DXFeedTimeSeria {
    setSymbol(symbol) {
        this.symbol = symbol;
        return this;
    }
    setEventTime(eventTime) {
        this.eventTime = eventTime;
        return this;
    }
    setEventFlags(eventFlags) {
        this.eventFlags = eventFlags;
        return this;
    }
    setIndex(index) {
        this.index = index;
        return this;
    }
    setTime(time) {
        this.time = time;
        return this;
    }
    setSequence(sequence) {
        this.sequence = sequence;
        return this;
    }
    setCount(count) {
        this.count = count;
        return this;
    }
    setOpen(open) {
        this.open = open;
        return this;
    }
    setHigh(high) {
        this.high = high;
        return this;
    }
    setLow(low) {
        this.low = low;
        return this;
    }
    setClose(close) {
        this.close = close;
        return this;
    }
    setVolume(volume) {
        this.volume = volume;
        return this;
    }
    setVwap(vwap) {
        this.vwap = vwap;
        return this;
    }
    setBidVolume(bidVolume) {
        this.bidVolume = bidVolume;
        return this;
    }
    setAskVolume(askVolume) {
        this.askVolume = askVolume;
        return this;
    }
    setImpVolatility(impVolatility) {
        this.impVolatility = impVolatility;
        return this;
    }
    setOpenInterest(openInterest) {
        this.openInterest = openInterest;
        return this;
    }
    build() {
        return this;
    }
}
module.exports = DXFeedTimeSeriaBuilder
