// @gehtsoft/traduAPI-node/lib/DxFeedReceiveQuotesCommand.js
'use strict';
let DxFeedReceiveQuotesPublisher = require('./DxFeedReceiveQuotesPublisher.js');
let LogManager = require('./LogManager.js');
let DxFeedReceiveQuotesDataReader = require('./DxFeedReceiveQuotesDataReader.js');
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
class DxFeedReceiveQuotesCommand {
    socketCommunicator;
    logger;
    socketReceiveMessageListener;
    dxFeedReceiveQuotesPublisher;
    constructor() {
        this.socketCommunicator = null;
        this.socketReceiveMessageListener = new DxFeedReceiveQuotesCommand.SocketReceiveMessageListener(this);
        this.dxFeedReceiveQuotesPublisher = new DxFeedReceiveQuotesPublisher();
        this.logger = LogManager.getLogger();
    }
    processMessage(json) {
        let reader, data, ex;
        reader = new DxFeedReceiveQuotesDataReader();
        try {
            data = reader.read(json);
            if (data.length > 0) {
                this.dxFeedReceiveQuotesPublisher.notifyReceivedQuotes(data);
            }
        } catch(exTemp1) {
            ex = stdlib.ensureException(exTemp1);
            this.logger.error("can not process DxFeedReceiveQuotesCommand response: " + ex.getMessage());
        }
    }
    subscribe(listener) {
        this.dxFeedReceiveQuotesPublisher.subscribe(listener);
    }
    unsubscribe(listener) {
        this.dxFeedReceiveQuotesPublisher.unsubscribe(listener);
    }
    stop() {
        this.socketCommunicator.unsubscribeJsonReceive(this.socketReceiveMessageListener);
    }
    onChange(state) {
        if (state.isOpen() && !state.hasError()) {
            this.socketCommunicator.subscribeJsonReceive(this.socketReceiveMessageListener);
        }
    }
    static create(socketCommunicator) {
        let dxFeedReceiveQuotesCommand;
        dxFeedReceiveQuotesCommand = new DxFeedReceiveQuotesCommand();
        dxFeedReceiveQuotesCommand.socketCommunicator = socketCommunicator;
        return dxFeedReceiveQuotesCommand;
    }
}
module.exports = DxFeedReceiveQuotesCommand
DxFeedReceiveQuotesCommand.SocketReceiveMessageListener = class {
    _owner_;
    constructor(_owner_) {
        this._owner_ = _owner_;
    }
    onReceive(json) {
        this._owner_.processMessage(json);
    }
}
