// @gehtsoft/traduAPI-node/lib/EntryOrderRequestBuilder.js
'use strict';
let EntryOrderRequest = require('./EntryOrderRequest.js');
//@@DocInclude
//@@DocBrief("The builder for creating instances of the EntryOrderRequest class, which is used for creating orders in a trading platform")
class EntryOrderRequestBuilder extends EntryOrderRequest {
    setAccountId(accountId) {
        this.accountId = accountId;
        return this;
    }
    setOfferId(offerId) {
        this.offerId = offerId;
        return this;
    }
    setAmount(amount) {
        this.amount = amount;
        return this;
    }
    setBuySell(buysell) {
        this.buysell = buysell;
        return this;
    }
    setRate(rate) {
        this.rate = rate;
        return this;
    }
    setStopRate(stopRate) {
        this.isStopOrderSetField = true;
        this.stopRate = stopRate;
        return this;
    }
    setLimitRate(limitRate) {
        this.isLimitOrderSetField = true;
        this.limitRate = limitRate;
        return this;
    }
    setStopPips(stopPips) {
        this.isStopOrderSetField = true;
        this.stopPips = stopPips;
        return this;
    }
    setLimitPips(limitPips) {
        this.isLimitOrderSetField = true;
        this.limitPips = limitPips;
        return this;
    }
    setRateRange(rateRange) {
        this.rateRange = rateRange;
        return this;
    }
    enableTrailingStop(type, stepInPips) {
        this.trailingStopType = type;
        this.trailingStopStep = stepInPips;
        return this;
    }
    setTimeInForce(timeInForce) {
        this.timeInForce = timeInForce;
        return this;
    }
    setExpirationDate(expirationDate) {
        this.expirationDate = expirationDate;
        return this;
    }
    setCustomId(customId) {
        this.customId = customId;
        return this;
    }
    build() {
        return this;
    }
}
module.exports = EntryOrderRequestBuilder
