// @gehtsoft/traduAPI-node/lib/GetPriceHistoryMessageBuilder.js
'use strict';
let GetPriceHistoryMessage = require('./GetPriceHistoryMessage.js');
let PriceHistoryList = require('./PriceHistoryList.js');
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
class GetPriceHistoryMessageBuilder extends GetPriceHistoryMessage {
    constructor() {
        super();
        this.priceHistoryList = new PriceHistoryList();
    }
    addPriceHistory(offer) {
        this.priceHistoryList.add(offer);
    }
    addPriceHistoryList(src) {
        let i;
        if (src == null) {
            stdlib.print("src == nil");
        }
        if (src != null) {
            for (i = 0; i <= src.length() - 1; i += 1) {
                this.priceHistoryList.add(src.get(i));
            }
        }
    }
    setStream(stream) {
        this.stream = stream;
    }
    setTimeframe(timeframe) {
        this.timeframe = timeframe;
    }
    setSymbol(symbol) {
        this.symbol = symbol;
    }
    setRequestId(requestId) {
        this.requestId = requestId;
    }
    setError(error) {
        this.error = error;
    }
    build() {
        return this;
    }
}
module.exports = GetPriceHistoryMessageBuilder
