// @gehtsoft/traduAPI-node/lib/GetTemporarySessionBySsoTokenCommandBuilder.js
'use strict';
let GetTemporarySessionBySsoTokenCommand = require('./GetTemporarySessionBySsoTokenCommand.js');
//@@DocBrief("The builder of the GetTemporarySessionBySsoTokenCommand class")
class GetTemporarySessionBySsoTokenCommandBuilder extends GetTemporarySessionBySsoTokenCommand {
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setSsoToken(ssoToken) {
        this.ssoToken = ssoToken;
    }
    setSsoTokenProvider(ssoTokenProvider) {
        this.ssoTokenProvider = ssoTokenProvider;
    }
    setLogin(login) {
        this.login = login;
    }
    setApplicationName(applicationName) {
        this.applicationName = applicationName;
    }
    setAgent(agent) {
        this.agent = agent;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setTerminal(terminal) {
        this.terminal = terminal;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setConfigFlags(configFlags) {
        this.configFlags = configFlags;
    }
    build() {
        return this;
    }
}
module.exports = GetTemporarySessionBySsoTokenCommandBuilder
