// @gehtsoft/traduAPI-node/lib/GetTemporaryPriceSessionBySsoTokenCommandBuilder.js
'use strict';
let GetTemporaryPriceSessionBySsoTokenCommand = require('./GetTemporaryPriceSessionBySsoTokenCommand.js');
//@@DocBrief("The builder of the GetTemporaryPriceSessionBySsoTokenCommand class")
class GetTemporaryPriceSessionBySsoTokenCommandBuilder extends GetTemporaryPriceSessionBySsoTokenCommand {
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setSsoToken(ssoToken) {
        this.ssoToken = ssoToken;
    }
    setLogin(login) {
        this.login = login;
    }
    setApplicationName(applicationName) {
        this.applicationName = applicationName;
    }
    setAgent(agent) {
        this.agent = agent;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setTerminal(terminal) {
        this.terminal = terminal;
    }
    setPriceTerminal(priceTerminal) {
        this.priceTerminal = priceTerminal;
    }
    setTradingSession(tradingSession) {
        this.tradingSession = tradingSession;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    build() {
        return this;
    }
}
module.exports = GetTemporaryPriceSessionBySsoTokenCommandBuilder
