// @gehtsoft/traduAPI-node/lib/GetTradingSessionCommandBuilder.js
'use strict';
let GetTradingSessionCommand = require('./GetTradingSessionCommand.js');
//@@DocBrief("The builder of the GetTemporarySessionCommand class")
class GetTradingSessionCommandBuilder extends GetTradingSessionCommand {
    setPeriodInMilliseconds(periodInMilliseconds) {
        this.periodInMilliseconds = periodInMilliseconds;
    }
    setTimeoutInSeconds(timeoutInSeconds) {
        this.timeoutInSeconds = timeoutInSeconds;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setTemporarySession(temporarySession) {
        this.temporarySession = temporarySession;
    }
    setLogin(login) {
        this.login = login;
    }
    setApplicationName(applicationName) {
        this.applicationName = applicationName;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setTerminal(terminal) {
        this.terminal = terminal;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    build() {
        return this;
    }
}
module.exports = GetTradingSessionCommandBuilder
