// @gehtsoft/traduAPI-node/lib/GetTemporarySessionCommandBuilder.js
'use strict';
let GetTemporarySessionCommand = require('./GetTemporarySessionCommand.js');
//@@DocBrief("The builder of the GetTemporarySessionCommand class")
class GetTemporarySessionCommandBuilder extends GetTemporarySessionCommand {
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setSaltedPassword(saltedPassword) {
        this.saltedPassword = saltedPassword;
    }
    setLogin(login) {
        this.login = login;
    }
    setApplicationName(applicationName) {
        this.applicationName = applicationName;
    }
    setAgent(agent) {
        this.agent = agent;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setTerminal(terminal) {
        this.terminal = terminal;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setConfigFlags(configFlags) {
        this.configFlags = configFlags;
    }
    build() {
        return this;
    }
}
module.exports = GetTemporarySessionCommandBuilder
