// @gehtsoft/traduAPI-node/lib/GetTemporaryPriceSessionCommandBuilder.js
'use strict';
let GetTemporaryPriceSessionCommand = require('./GetTemporaryPriceSessionCommand.js');
//@@DocBrief("The builder of the GetTemporaryPriceSessionCommand class")
class GetTemporaryPriceSessionCommandBuilder extends GetTemporaryPriceSessionCommand {
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setSaltedPassword(saltedPassword) {
        this.saltedPassword = saltedPassword;
    }
    setLogin(login) {
        this.login = login;
    }
    setApplicationName(applicationName) {
        this.applicationName = applicationName;
    }
    setAgent(agent) {
        this.agent = agent;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setTerminal(terminal) {
        this.terminal = terminal;
    }
    setTradingSession(tradingSession) {
        this.tradingSession = tradingSession;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    build() {
        return this;
    }
}
module.exports = GetTemporaryPriceSessionCommandBuilder
