// @gehtsoft/traduAPI-node/lib/InstrumentDescriptorsTransform.js
'use strict';
let list = require('@gehtsoft/LuaxStdlib-node').list;
let InstrumentDescriptorFindingResult = require('./InstrumentDescriptorFindingResult.js');
class InstrumentDescriptorsTransform {
    descriptorsStorage;
    instrumentDescriptorsSeparator;
    static create(descriptorsStorage, instrumentDescriptorsSeparator) {
        let result;
        result = new InstrumentDescriptorsTransform();
        result.descriptorsStorage = descriptorsStorage;
        result.instrumentDescriptorsSeparator = instrumentDescriptorsSeparator;
        return result;
    }
    selectPdasInstruments(symbols) {
        return this.modify(symbols);
    }
    modify(symbols) {
        let descriptors, result, i, error;
        descriptors = new list();
        error = this.addDescriptors(symbols, descriptors);
        result = new Array(descriptors.length());
        for (i = 0; i <= descriptors.length() - 1; i += 1) {
            result[i] = descriptors.get(i);
        }
        return InstrumentDescriptorFindingResult.create(result, error);
    }
    addDescriptors(symbols, descriptors) {
        let instrumentDescriptors, descriptor, i, error;
        error = null;
        for (i = 0; i <= symbols.length - 1; i += 1) {
            descriptor = this.descriptorsStorage.getInstrumentDescriptorBySymbol(symbols[i]);
            instrumentDescriptors = this.instrumentDescriptorsSeparator.getPdasDescriptors([descriptor]);
            if (instrumentDescriptors == null || instrumentDescriptors.length == 0) {
                continue;
            }
            if (descriptor != null) {
                descriptors.add(descriptor);
            }
            else {
                error = this.addError(error, symbols[i]);
            }
        }
        return error;
    }
    addError(error, symbol) {
        let result;
        result = error;
        if (result == null) {
            result = "Not found instruments for symbols: " + symbol;
        }
        else {
            result = result + ", " + symbol;
        }
        return result;
    }
}
module.exports = InstrumentDescriptorsTransform
