// @gehtsoft/traduAPI-node/lib/InstrumentUpdate.js
'use strict';
//@@DocBrief("The class provides access to instrument update information")
class InstrumentUpdate {
    instrument;
    mIsOfferIdChanged;
    mIsSymbolChanged;
    mIsContractCurrencyChanged;
    mIsDigitsChanged;
    mIsPointSizeChanged;
    mIsInstrumentTypeChanged;
    mIsTradingStatusChanged;
    mIsContractMultiplierChanged;
    mIsSellInterestChanged;
    mIsBuyInterestChanged;
    mIsSubscriptionStatusChanged;
    mIsBaseUnitSizeChanged;
    mIsAskAdjustmentChanged;
    mIsBidAdjustmentChanged;
    mIsDividendBuyChanged;
    mIsDividendSellChanged;
    mIsMinQuantityChanged;
    mIsMaxQuantityChanged;
    mIsSortOrderChanged;
    mIsPriceStreamIdChanged;
    mIsConditionDistStopChanged;
    mIsConditionDistLimitChanged;
    mIsConditionDistEntryStopChanged;
    mIsConditionDistEntryLimitChanged;
    mIsMmrChanged;
    mIsMmrByCurrencyChanged;
    constructor() {
        this.mIsOfferIdChanged = false;
        this.mIsSymbolChanged = false;
        this.mIsContractCurrencyChanged = false;
        this.mIsDigitsChanged = false;
        this.mIsPointSizeChanged = false;
        this.mIsInstrumentTypeChanged = false;
        this.mIsTradingStatusChanged = false;
        this.mIsContractMultiplierChanged = false;
        this.mIsSellInterestChanged = false;
        this.mIsBuyInterestChanged = false;
        this.mIsSubscriptionStatusChanged = false;
        this.mIsBaseUnitSizeChanged = false;
        this.mIsDividendSellChanged = false;
        this.mIsDividendBuyChanged = false;
        this.mIsBidAdjustmentChanged = false;
        this.mIsAskAdjustmentChanged = false;
        this.mIsMinQuantityChanged = false;
        this.mIsMaxQuantityChanged = false;
        this.mIsSortOrderChanged = false;
        this.mIsPriceStreamIdChanged = false;
        this.mIsConditionDistStopChanged = false;
        this.mIsConditionDistLimitChanged = false;
        this.mIsConditionDistEntryStopChanged = false;
        this.mIsConditionDistEntryLimitChanged = false;
        this.mIsMmrByCurrencyChanged = false;
        this.mIsMmrChanged = false;
    }
    getInstrument() {
        return this.instrument;
    }
    isOfferIdChanged() {
        return this.mIsOfferIdChanged;
    }
    isSymbolChanged() {
        return this.mIsSymbolChanged;
    }
    isContractCurrencyChanged() {
        return this.mIsContractCurrencyChanged;
    }
    isDigitsChanged() {
        return this.mIsDigitsChanged;
    }
    isPointSizeChanged() {
        return this.mIsPointSizeChanged;
    }
    isInstrumentTypeChanged() {
        return this.mIsInstrumentTypeChanged;
    }
    isTradingStatusChanged() {
        return this.mIsTradingStatusChanged;
    }
    isContractMultiplierChanged() {
        return this.mIsContractMultiplierChanged;
    }
    isSellInterestChanged() {
        return this.mIsSellInterestChanged;
    }
    isBuyInterestChanged() {
        return this.mIsBuyInterestChanged;
    }
    isSubscriptionStatusChanged() {
        return this.mIsSubscriptionStatusChanged;
    }
    isBaseUnitSizeChanged() {
        return this.mIsBaseUnitSizeChanged;
    }
    isAskAdjustmentChanged() {
        return this.mIsAskAdjustmentChanged;
    }
    isBidAdjustmentChanged() {
        return this.mIsBidAdjustmentChanged;
    }
    isDividendBuyChanged() {
        return this.mIsDividendBuyChanged;
    }
    isDividendSellChanged() {
        return this.mIsDividendSellChanged;
    }
    isMinQuantityChanged() {
        return this.mIsMinQuantityChanged;
    }
    isMaxQuantityChanged() {
        return this.mIsMaxQuantityChanged;
    }
    isSortOrderChanged() {
        return this.mIsSortOrderChanged;
    }
    isPriceStreamIdChanged() {
        return this.mIsPriceStreamIdChanged;
    }
    isConditionDistStopChanged() {
        return this.mIsConditionDistStopChanged;
    }
    isConditionDistLimitChanged() {
        return this.mIsConditionDistLimitChanged;
    }
    isConditionDistEntryStopChanged() {
        return this.mIsConditionDistEntryStopChanged;
    }
    isConditionDistEntryLimitChanged() {
        return this.mIsConditionDistEntryLimitChanged;
    }
    isMmrChanged() {
        return this.mIsMmrChanged;
    }
    isMmrByCurrencyChanged() {
        return this.mIsMmrByCurrencyChanged;
    }
    mergeToInstrument(mergeTo) {
        this.mIsOfferIdChanged = this.instrument.getOfferId() != mergeTo.getOfferId();
        this.mIsSymbolChanged = this.instrument.getSymbol() != mergeTo.getSymbol();
        this.mIsContractCurrencyChanged = this.instrument.getContractCurrency() != mergeTo.getContractCurrency();
        this.mIsDigitsChanged = this.instrument.getDigits() != mergeTo.getDigits();
        this.mIsPointSizeChanged = this.instrument.getPointSize() != mergeTo.getPointSize();
        this.mIsInstrumentTypeChanged = this.instrument.getInstrumentType() != mergeTo.getInstrumentType();
        this.mIsTradingStatusChanged = this.instrument.getTradingStatus() != mergeTo.getTradingStatus();
        this.mIsContractMultiplierChanged = this.instrument.getContractMultiplier() != mergeTo.getContractMultiplier();
        this.mIsSellInterestChanged = this.instrument.getSellInterest() != mergeTo.getSellInterest();
        this.mIsBuyInterestChanged = this.instrument.getBuyInterest() != mergeTo.getBuyInterest();
        this.mIsSubscriptionStatusChanged = this.instrument.getSubscriptionStatus() != mergeTo.getSubscriptionStatus();
        this.mIsDividendBuyChanged = this.instrument.getDividendBuy() != mergeTo.getDividendBuy();
        this.mIsDividendSellChanged = this.instrument.getDividendSell() != mergeTo.getDividendSell();
        this.mIsBaseUnitSizeChanged = this.instrument.getBaseUnitSize() != mergeTo.getBaseUnitSize();
        this.mIsMinQuantityChanged = this.instrument.getMinQuantity() != mergeTo.getMinQuantity();
        this.mIsMaxQuantityChanged = this.instrument.getMaxQuantity() != mergeTo.getMaxQuantity();
        this.mIsSortOrderChanged = this.instrument.getSortOrder() != mergeTo.getSortOrder();
        this.mIsPriceStreamIdChanged = this.instrument.getPriceStreamId() != mergeTo.getPriceStreamId();
        this.mIsConditionDistStopChanged = this.instrument.getConditionDistStop() != mergeTo.getConditionDistStop();
        this.mIsConditionDistLimitChanged = this.instrument.getConditionDistLimit() != mergeTo.getConditionDistLimit();
        this.mIsConditionDistEntryStopChanged = this.instrument.getConditionDistEntryStop() != mergeTo.getConditionDistEntryStop();
        this.mIsConditionDistEntryLimitChanged = this.instrument.getConditionDistEntryLimit() != mergeTo.getConditionDistEntryLimit();
        this.mIsAskAdjustmentChanged = this.instrument.getAskAdjustment() != mergeTo.getAskAdjustment();
        this.mIsBidAdjustmentChanged = this.instrument.getBidAdjustment() != mergeTo.getBidAdjustment();
        this.mIsMmrChanged = this.instrument.getMMR() != mergeTo.getMMR();
        this.mIsMmrByCurrencyChanged = true;
    }
}
module.exports = InstrumentUpdate
