// @gehtsoft/traduAPI-node/lib/JWT.js
'use strict';
class JWT {
    tradingSessionId;
    tradingSessionSubId;
    sessionId;
    userId;
    userKind;
    login;
    ssoToken;
    constructor() {
        this.tradingSessionId = "";
        this.tradingSessionSubId = "";
        this.sessionId = "";
        this.userId = "";
        this.userKind = "";
        this.login = "";
        this.ssoToken = "";
    }
    static create(tradingSessionId, tradingSessionSubId, sessionId, userId, userKind, login, ssoToken) {
        let result;
        result = new JWT();
        result.tradingSessionId = tradingSessionId;
        result.tradingSessionSubId = tradingSessionSubId;
        result.sessionId = sessionId;
        result.userId = userId;
        result.userKind = userKind;
        result.login = login;
        result.ssoToken = ssoToken;
        return result;
    }
    getLogin() {
        return this.login;
    }
    getUserKind() {
        return this.userKind;
    }
    getUserId() {
        return this.userId;
    }
    getTradingSessionId() {
        return this.tradingSessionId;
    }
    getTradingSessionSubId() {
        return this.tradingSessionSubId;
    }
    getSessionId() {
        return this.sessionId;
    }
    getSsoToken() {
        return this.ssoToken;
    }
}
module.exports = JWT
