// @gehtsoft/traduAPI-node/lib/KeepAlivePriceSessionCommand.js
'use strict';
let KeepAliveCommand = require('./KeepAliveCommand.js');
let KeepAlivePriceSessionTransportMessageFactory = require('./KeepAlivePriceSessionTransportMessageFactory.js');
let MessageType = require('./MessageType.js');
let StringUtil = require('./StringUtil.js');
//@@DocBrief("The command for sending a 'keep alive' message")
class KeepAlivePriceSessionCommand extends KeepAliveCommand {
    constructor() {
        super();
        this.checkRequest = new KeepAlivePriceSessionCommand.CheckKeepAliveRequest(this);
    }
    getSession() {
        return this.session;
    }
    getCommandName() {
        return "KeepAlivePriceSessionCommand(" + this.priceTerminal.getSubId() + ")";
    }
    createKeepAliveTransportMessageFactory() {
        return new KeepAlivePriceSessionTransportMessageFactory();
    }
    createMessageType() {
        return MessageType.KeepAlivePrices;
    }
}
module.exports = KeepAlivePriceSessionCommand
KeepAlivePriceSessionCommand.CheckKeepAliveRequest = class {
    _owner_;
    constructor(_owner_) {
        this._owner_ = _owner_;
    }
    isMyRequest(requestId) {
        return StringUtil.getSessionId(requestId) == this._owner_.getSession().getSessionId();
    }
}
