// @gehtsoft/traduAPI-node/lib/KeepAlivePriceSessionCommandBuilder.js
'use strict';
let KeepAlivePriceSessionCommand = require('./KeepAlivePriceSessionCommand.js');
//@@DocBrief("The builder of the KeepAlivePriceSessionCommand class")
class KeepAlivePriceSessionCommandBuilder extends KeepAlivePriceSessionCommand {
    setPeriodInMilliseconds(periodInMilliseconds) {
        this.periodInMilliseconds = periodInMilliseconds;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setSession(session) {
        this.session = session;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setConnectionParameters(connectionParameters) {
        this.connectionParameters = connectionParameters;
    }
    setPriceTerminal(priceTerminal) {
        this.priceTerminal = priceTerminal;
    }
    build() {
        this.createSchedulers();
        return this;
    }
}
module.exports = KeepAlivePriceSessionCommandBuilder
