// @gehtsoft/traduAPI-node/lib/KeepAliveCommandBuilder.js
'use strict';
let KeepAliveCommand = require('./KeepAliveCommand.js');
//@@DocBrief("The builder of the KeepAliveCommand class")
class KeepAliveCommandBuilder extends KeepAliveCommand {
    setPeriodInMilliseconds(periodInMilliseconds) {
        this.periodInMilliseconds = periodInMilliseconds;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setSession(session) {
        super.setSession(session);
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setConnectionParameters(connectionParameters) {
        this.connectionParameters = connectionParameters;
    }
    build() {
        this.createSchedulers();
        return this;
    }
}
module.exports = KeepAliveCommandBuilder
