// @gehtsoft/traduAPI-node/lib/LeverageProfile.js
'use strict';
//@@DocBrief("The class provides access to account leverage profile information")
class LeverageProfile {
    accountId;
    alpId;
    offerID;
    usedMargin;
    usedMarginAware;
    usedMarginEntry;
    changeNo;
    progressiveMargin;
    calcPath;
    marginType;
    marginCoeff;
    marginAwareCoeff;
    marginEntryCoeff;
    constructor() {
        this.accountId = "";
        this.alpId = "";
        this.offerID = "";
        this.usedMargin = 0.0;
        this.usedMarginAware = 0.0;
        this.usedMarginEntry = 0.0;
        this.changeNo = 0;
        this.progressiveMargin = "";
        this.calcPath = "";
        this.marginType = "V";
        this.marginCoeff = 0.0;
        this.marginAwareCoeff = 0.0;
        this.marginEntryCoeff = 0.0;
    }
    getAccountID() {
        return this.accountId;
    }
    getAlpID() {
        return this.alpId;
    }
    getOfferID() {
        return this.offerID;
    }
    getUsedMargin() {
        return this.usedMargin;
    }
    getUsedMarginAware() {
        return this.usedMarginAware;
    }
    getUsedMarginEntry() {
        return this.usedMarginEntry;
    }
    getChangeNo() {
        return this.changeNo;
    }
    getProgressiveMargin() {
        return this.progressiveMargin;
    }
    getCalcPath() {
        return this.calcPath;
    }
    getMarginType() {
        return this.marginType;
    }
    getMarginCoeff() {
        return this.marginCoeff;
    }
    getMarginAwareCoeff() {
        return this.marginAwareCoeff;
    }
    getMarginEntryCoeff() {
        return this.marginEntryCoeff;
    }
}
module.exports = LeverageProfile
