// @gehtsoft/traduAPI-node/lib/LeverageProfileBuilder.js
'use strict';
let LeverageProfile = require('./LeverageProfile.js');
let variantCast = require('@gehtsoft/LuaxStdlib-node').variantCast;
//@@DocBrief("The builder of the LeverageProfile class")
class LeverageProfileBuilder extends LeverageProfile {
    setAccountId(value) {
        this.accountId = value;
    }
    setAlpId(value) {
        this.alpId = value;
    }
    setOfferId(value) {
        this.offerID = value;
    }
    setUsedMargin(value) {
        this.usedMargin = variantCast.castToReal(variantCast.fromString(value));
    }
    setUsedMarginAware(value) {
        this.usedMarginAware = variantCast.castToReal(variantCast.fromString(value));
    }
    setUsedMarginEntry(value) {
        this.usedMarginEntry = variantCast.castToReal(variantCast.fromString(value));
    }
    setChangeNo(value) {
        this.changeNo = variantCast.castToInt(variantCast.fromString(value));
    }
    setProgressiveMargin(value) {
        this.progressiveMargin = value;
    }
    setCalcPath(value) {
        this.calcPath = value;
    }
    setMarginType(value) {
        this.marginType = value;
    }
    setMarginCoeff(value) {
        this.marginCoeff = variantCast.castToReal(variantCast.fromString(value));
    }
    setMarginAwareCoeff(value) {
        this.marginAwareCoeff = variantCast.castToReal(variantCast.fromString(value));
    }
    setMarginEntryCoeff(value) {
        this.marginEntryCoeff = variantCast.castToReal(variantCast.fromString(value));
    }
    build() {
        return this;
    }
}
module.exports = LeverageProfileBuilder
