// @gehtsoft/traduAPI-node/lib/MarginDescriptor.js
'use strict';
class MarginDescriptor {
    usedMargin;
    usedMarginAware;
    usedMarginEntry;
    progressiveMargin;
    marginType;
    marginCoeff;
    marginAwareCoeff;
    marginEntryCoeff;
    getUsedMargin() {
        return this.usedMargin;
    }
    getUsedMarginAware() {
        return this.usedMarginAware;
    }
    getUsedMarginEntry() {
        return this.usedMarginEntry;
    }
    getProgressiveMargin() {
        return this.progressiveMargin;
    }
    getMarginType() {
        return this.marginType;
    }
    getMarginCoeff() {
        return this.marginCoeff;
    }
    getMarginAwareCoeff() {
        return this.marginAwareCoeff;
    }
    getMarginEntryCoeff() {
        return this.marginEntryCoeff;
    }
    static create(usedMargin, usedMarginAware, usedMarginEntry, progressiveMargin, marginType, marginCoeff, marginAwareCoeff, marginEntryCoeff) {
        let result;
        result = new MarginDescriptor();
        result.usedMargin = usedMargin;
        result.usedMarginAware = usedMarginAware;
        result.usedMarginEntry = usedMarginEntry;
        result.progressiveMargin = progressiveMargin;
        result.marginType = marginType;
        result.marginCoeff = marginCoeff;
        result.marginAwareCoeff = marginAwareCoeff;
        result.marginEntryCoeff = marginEntryCoeff;
        return result;
    }
}
module.exports = MarginDescriptor
