// @gehtsoft/traduAPI-node/lib/MarketOrderRequest.js
'use strict';
let OrderRequest = require('./OrderRequest.js');
let TrailingStopType = require('./TrailingStopType.js');
//@@DocInclude
//@@DocBrief("The class provides information for a market order request")
class MarketOrderRequest extends OrderRequest {
    stopPips;
    limitPips;
    stopRate;
    limitRate;
    isStopOrderSetField;
    isLimitOrderSetField;
    rateRange;
    rateRangeFilled;
    trailingStopType;
    trailingStopStep;
    constructor() {
        super();
        this.isStopOrderSetField = false;
        this.isLimitOrderSetField = false;
        this.rateRangeFilled = false;
        this.limitRate = 0.0;
        this.stopRate = 0.0;
        this.rateRange = 0.0;
        this.trailingStopType = TrailingStopType.None;
        this.trailingStopStep = 0;
        this.stopPips = 0.0;
        this.limitPips = 0.0;
    }
    getStopPips() {
        return this.stopPips;
    }
    getLimitPips() {
        return this.limitPips;
    }
    getStopRate() {
        return this.stopRate;
    }
    getLimitRate() {
        return this.limitRate;
    }
    isStopOrderSet() {
        return this.isStopOrderSetField;
    }
    isLimitOrderSet() {
        return this.isLimitOrderSetField;
    }
    getRateRange() {
        return this.rateRange;
    }
    isRateRangeFilled() {
        return this.rateRangeFilled;
    }
    isEnabledTrailingStop() {
        return this.trailingStopType != TrailingStopType.None;
    }
    getTrailingStopType() {
        return this.trailingStopType;
    }
    getTrailingStopStep() {
        return this.trailingStopStep;
    }
}
module.exports = MarketOrderRequest
