// @gehtsoft/traduAPI-node/lib/MarketOrderRequestBuilder.js
'use strict';
let MarketOrderRequest = require('./MarketOrderRequest.js');
//@@DocInclude
//@@DocBrief("The builder of the MarketOrderRequest class")
class MarketOrderRequestBuilder extends MarketOrderRequest {
    constructor() {
        super();
        this.customId = "";
    }
    setAccountId(accountId) {
        this.accountId = accountId;
        return this;
    }
    setOfferId(offerId) {
        this.offerId = offerId;
        return this;
    }
    setAmount(amount) {
        this.amount = amount;
        return this;
    }
    setBuySell(buysell) {
        this.buysell = buysell;
        return this;
    }
    setTimeInForce(timeInForce) {
        this.timeInForce = timeInForce;
        return this;
    }
    setExpirationDate(expirationDate) {
        this.expirationDate = expirationDate;
        return this;
    }
    setCustomId(customId) {
        this.customId = customId;
        return this;
    }
    setStopRate(rate) {
        this.isStopOrderSetField = true;
        this.stopRate = rate;
        return this;
    }
    setLimitRate(rate) {
        this.isLimitOrderSetField = true;
        this.limitRate = rate;
        return this;
    }
    setStopPips(stopPips) {
        this.isStopOrderSetField = true;
        this.stopPips = stopPips;
        return this;
    }
    setLimitPips(limitPips) {
        this.isLimitOrderSetField = true;
        this.limitPips = limitPips;
        return this;
    }
    setRateRange(rateRange) {
        this.rateRange = rateRange;
        this.rateRangeFilled = true;
        return this;
    }
    enableTrailingStop(type, stepInPips) {
        this.trailingStopType = type;
        this.trailingStopStep = stepInPips;
        return this;
    }
    build() {
        return this;
    }
}
module.exports = MarketOrderRequestBuilder
