// @gehtsoft/traduAPI-node/lib/OffersStorage.js
'use strict';
let string_map = require('@gehtsoft/LuaxStdlib-node').string_map;
//@@DocBrief("class provides access to a collection of offer information")
class OffersStorage {
    storage;
    constructor() {
        this.storage = new string_map();
    }
    getAllOffers() {
        let length, i, keys, offer, offers;
        keys = this.storage.keys();
        length = keys.length;
        offers = new Array(length);
        for (i = 0; i <= length - 1; i += 1) {
            offer = this.storage.get(keys[i]);
            offers[i] = offer;
        }
        return offers;
    }
    addOffer(offer) {
        this.storage.set(offer.getOfferId(), offer);
    }
    removeOfferById(offerId) {
        if (this.storage.contains(offerId)) {
            this.storage.remove(offerId);
        }
    }
    getOfferById(offerId) {
        if (!this.storage.contains(offerId)) {
            return null;
        }
        return this.storage.get(offerId);
    }
    clear() {
        this.storage.clear();
    }
    getCount() {
        return this.storage.length();
    }
    getOfferIds() {
        return this.storage.keys();
    }
}
module.exports = OffersStorage
