// @gehtsoft/traduAPI-node/lib/OfferSnapshotCommandBuilder.js
'use strict';
let OfferSnapshotCommand = require('./OfferSnapshotCommand.js');
class OfferSnapshotCommandBuilder extends OfferSnapshotCommand {
    setSessionStorage(sessionStorage) {
        this.sessionStorage = sessionStorage;
    }
    setTimeController(timeController) {
        this.timeController = timeController;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setRequestNumber(requestNumber) {
        this.requestNumber = requestNumber;
    }
    setInstruments(instruments) {
        this.instruments = instruments;
    }
    setCommandDescriptor(descriptor) {
        this.descriptor = descriptor;
    }
    setDxFeedNamesProvider(dxFeedNamesProvider) {
        this.dxFeedNamesProvider = dxFeedNamesProvider;
    }
    setCommandCallback(callback) {
        this.rawOfferSnapshotCommandCallback = callback;
    }
    build() {
        return this;
    }
}
module.exports = OfferSnapshotCommandBuilder
