// @gehtsoft/traduAPI-node/lib/OfferUpdate.js
'use strict';
//@@DocBrief("The class provides access to offer update information")
class OfferUpdate {
    offer;
    isQuoteIdChangedField;
    isBidChangedField;
    isAskChangedField;
    isBidTradableChangedField;
    isAskTradableChangedField;
    isHighChangedField;
    isLowChangedField;
    isVolumeChangedField;
    isTimeChangedField;
    getOffer() {
        return this.offer;
    }
    isQuoteIdChanged() {
        return this.isQuoteIdChangedField;
    }
    isBidChanged() {
        return this.isBidChangedField;
    }
    isAskChanged() {
        return this.isAskChangedField;
    }
    isBidTradableChanged() {
        return this.isBidTradableChangedField;
    }
    isAskTradableChanged() {
        return this.isAskTradableChangedField;
    }
    isHighChanged() {
        return this.isHighChangedField;
    }
    isLowChanged() {
        return this.isLowChangedField;
    }
    isVolumeChanged() {
        return this.isVolumeChangedField;
    }
    isTimeChanged() {
        return this.isTimeChangedField;
    }
}
module.exports = OfferUpdate
