// @gehtsoft/traduAPI-node/lib/Offer.js
'use strict';
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
//@@DocInclude
//@@DocBrief("The class provides access to offer information of a trading instrument, its current prices, high and low prices for the trading day")
class Offer {
    bid;
    ask;
    high;
    low;
    offerId;
    bidTradable;
    askTradable;
    quoteId;
    volume;
    time;
    constructor() {
        this.offerId = "";
        this.quoteId = "";
        this.bidTradable = "";
        this.askTradable = "";
        this.time = stdlib.mkdate(1900, 1, 1);
        this.bid = 0.0;
        this.ask = 0.0;
        this.high = 0.0;
        this.low = 0.0;
        this.volume = 0;
    }
    getBid() {
        return this.bid;
    }
    getAsk() {
        return this.ask;
    }
    getHigh() {
        return this.high;
    }
    getLow() {
        return this.low;
    }
    getOfferId() {
        return this.offerId;
    }
    getBidTradable() {
        return this.bidTradable;
    }
    getAskTradable() {
        return this.askTradable;
    }
    getQuoteId() {
        return this.quoteId;
    }
    getVolume() {
        return this.volume;
    }
    getTime() {
        return this.time;
    }
}
module.exports = Offer
