// @gehtsoft/traduAPI-node/lib/OpenPosition.js
'use strict';
let RawOpenPosition = require('./RawOpenPosition.js');
//@@DocInclude
//@@DocBrief("The class provides access to information about an open position")
class OpenPosition extends RawOpenPosition {
    pl;
    plPips;
    grossPL;
    netPL;
    close;
    stop;
    limit;
    stopOrderID;
    limitOrderID;
    usedMargin;
    stopMove;
    constructor() {
        super();
        this.pl = 0;
        this.plPips = 0;
        this.netPL = 0.0;
        this.grossPL = 0;
        this.close = 0;
        this.stop = 0;
        this.limit = 0;
        this.usedMargin = 0;
        this.stopMove = 0.0;
        this.limitOrderID = null;
        this.stopOrderID = null;
    }
    getPL() {
        return this.pl;
    }
    getPLPips() {
        return this.plPips;
    }
    getGrossPL() {
        return this.grossPL;
    }
    getNetPL() {
        return this.netPL;
    }
    getCloseRate() {
        return this.close;
    }
    getStopRate() {
        return this.stop;
    }
    getLimitRate() {
        return this.limit;
    }
    getStopOrderID() {
        return this.stopOrderID;
    }
    getLimitOrderID() {
        return this.limitOrderID;
    }
    getUsedMargin() {
        return this.usedMargin;
    }
    getStopMove() {
        return this.stopMove;
    }
}
module.exports = OpenPosition
