// @gehtsoft/traduAPI-node/lib/OpenPositionBuilder.js
'use strict';
let OpenPosition = require('./OpenPosition.js');
//@@DocBrief("The builder of the OpenPosition class")
class OpenPositionBuilder extends OpenPosition {
    setTradeID(value) {
        this.tradeID = value;
    }
    setAccountId(value) {
        this.accountId = value;
    }
    setAccountName(value) {
        this.accountName = value;
    }
    setAccountKind(value) {
        this.accountKind = value;
    }
    setOfferId(value) {
        this.offerId = value;
    }
    setAmount(value) {
        this.amount = value;
    }
    setBuySell(value) {
        this.buySell = value;
    }
    setOpenRate(value) {
        this.openRate = value;
    }
    setOpenTime(value) {
        this.openTime = value;
    }
    setOpenQuoteId(value) {
        this.openQuoteId = value;
    }
    setOpenOrderId(value) {
        this.openOrderId = value;
    }
    setOpenOrderReqId(value) {
        this.openOrderReqId = value;
    }
    setOpenOrderRequestTXT(value) {
        this.openOrderRequestTXT = value;
    }
    setCommission(value) {
        this.commission = value;
    }
    setRolloverInterest(value) {
        this.rolloverInterest = value;
    }
    setTradeIdOrigin(value) {
        this.tradeIdOrigin = value;
    }
    setValueDate(value) {
        this.valueDate = value;
    }
    setParties(value) {
        this.parties = value;
    }
    setDividends(value) {
        this.dividends = value;
    }
    setPL(value) {
        this.pl = value;
    }
    setPLPips(value) {
        this.plPips = value;
    }
    setGrossPL(value) {
        this.grossPL = value;
    }
    setNetPL(value) {
        this.netPL = value;
    }
    setCloseRate(value) {
        this.close = value;
    }
    setStopRate(value) {
        this.stop = value;
    }
    setLimitRate(value) {
        this.limit = value;
    }
    setStopOrderID(value) {
        this.stopOrderID = value;
    }
    setLimitOrderID(value) {
        this.limitOrderID = value;
    }
    setUsedMargin(value) {
        this.usedMargin = value;
    }
    setStopMove(stopMove) {
        this.stopMove = stopMove;
    }
    build() {
        return this;
    }
}
module.exports = OpenPositionBuilder
