// @gehtsoft/traduAPI-node/lib/OpenPositionUpdate.js
'use strict';
//@@DocBrief("The class provides access to open position update information")
class OpenPositionUpdate {
    openPosition;
    tradeIdChanged;
    accountIdChanged;
    accountNameChanged;
    accountKindChanged;
    offerIdChanged;
    amountChanged;
    buySellChanged;
    openRateChanged;
    openTimeChanged;
    openQuoteIdChanged;
    openOrderIdChanged;
    openOrderReqIdChanged;
    openOrderRequestTXTChanged;
    commissionChanged;
    rolloverInterestChanged;
    tradeIdOriginChanged;
    valueDateChanged;
    partiesChanged;
    dividendsChanged;
    getOpenPosition() {
        return this.openPosition;
    }
    isTradeIdChanged() {
        return this.tradeIdChanged;
    }
    isAccountIdChanged() {
        return this.accountIdChanged;
    }
    isAccountNameChanged() {
        return this.accountNameChanged;
    }
    isAccountKindChanged() {
        return this.accountKindChanged;
    }
    isOfferIdChanged() {
        return this.offerIdChanged;
    }
    isAmountChanged() {
        return this.amountChanged;
    }
    isBuySellChanged() {
        return this.buySellChanged;
    }
    isOpenRateChanged() {
        return this.openRateChanged;
    }
    isOpenTimeChanged() {
        return this.openTimeChanged;
    }
    isOpenQuoteIdChanged() {
        return this.openQuoteIdChanged;
    }
    isOpenOrderIdChanged() {
        return this.openOrderIdChanged;
    }
    isOpenOrderReqIdChanged() {
        return this.openOrderReqIdChanged;
    }
    isOpenOrderRequestTXTChanged() {
        return this.openOrderRequestTXTChanged;
    }
    isCommissionChanged() {
        return this.commissionChanged;
    }
    isRolloverInterestChanged() {
        return this.rolloverInterestChanged;
    }
    isTradeIdOriginChanged() {
        return this.tradeIdOriginChanged;
    }
    isValueDateChanged() {
        return this.valueDateChanged;
    }
    isPartiesChanged() {
        return this.partiesChanged;
    }
    isDividendsChanged() {
        return this.dividendsChanged;
    }
}
module.exports = OpenPositionUpdate
