// @gehtsoft/traduAPI-node/lib/OpenPositionCalculatedFields.js
'use strict';
class OpenPositionCalculatedFields {
    pl;
    plPips;
    grossPl;
    closeRate;
    stopRate;
    stopOrderId;
    limitOrderId;
    limitRate;
    usedMargin;
    usedMargin3;
    netPL;
    stopMove;
    constructor() {
        this.pl = 0.0;
        this.plPips = 0.0;
        this.grossPl = 0.0;
        this.closeRate = 0.0;
        this.stopRate = 0.0;
        this.limitRate = 0.0;
        this.usedMargin = 0.0;
        this.usedMargin3 = 0.0;
        this.stopMove = 0.0;
        this.netPL = 0.0;
        this.stopOrderId = "";
        this.limitOrderId = "";
    }
    getPL() {
        return this.pl;
    }
    getPLPips() {
        return this.plPips;
    }
    getGrossPL() {
        return this.grossPl;
    }
    getNetPL() {
        return this.netPL;
    }
    getCloseRate() {
        return this.closeRate;
    }
    getStopRate() {
        return this.stopRate;
    }
    getLimitRate() {
        return this.limitRate;
    }
    getUsedMargin() {
        return this.usedMargin;
    }
    getUsedMargin3() {
        return this.usedMargin3;
    }
    getStopOrderId() {
        return this.stopOrderId;
    }
    getLimitOrderId() {
        return this.limitOrderId;
    }
    getStopMove() {
        return this.stopMove;
    }
}
module.exports = OpenPositionCalculatedFields
