// @gehtsoft/traduAPI-node/lib/OpenPositionCalculatedFieldsBuilder.js
'use strict';
let OpenPositionCalculatedFields = require('./OpenPositionCalculatedFields.js');
class OpenPositionCalculatedFieldsBuilder extends OpenPositionCalculatedFields {
    setPL(value) {
        this.pl = value;
    }
    setPLPips(value) {
        this.plPips = value;
    }
    setGrossPL(value) {
        this.grossPl = value;
    }
    setNetPL(value) {
        this.netPL = value;
    }
    setCloseRate(value) {
        this.closeRate = value;
    }
    setStopRate(value) {
        this.stopRate = value;
    }
    setLimitRate(value) {
        this.limitRate = value;
    }
    setUsedMargin(value) {
        this.usedMargin = value;
    }
    setUsedMargin3(value) {
        this.usedMargin3 = value;
    }
    setStopOrderId(value) {
        this.stopOrderId = value;
    }
    setLimitOrderId(value) {
        this.limitOrderId = value;
    }
    setStopMove(stopMove) {
        this.stopMove = stopMove;
    }
    build() {
        return this;
    }
}
module.exports = OpenPositionCalculatedFieldsBuilder
