// @gehtsoft/traduAPI-node/lib/OpenPositionUpdateBuilder.js
'use strict';
let OpenPositionUpdate = require('./OpenPositionUpdate.js');
let OpenPositionBuilder = require('./OpenPositionBuilder.js');
let FieldChangeChecker = require('./FieldChangeChecker.js');
let variantCast = require('@gehtsoft/LuaxStdlib-node').variantCast;
let DateTimeParser = require('./DateTimeParser.js');
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
//@@DocBrief("The builder of the OpenPosition class")
class OpenPositionUpdateBuilder extends OpenPositionUpdate {
    openPositionBuilder;
    constructor() {
        super();
        this.openPositionBuilder = new OpenPositionBuilder();
    }
    setTradeID(value) {
        this.tradeIdChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setTradeID(value);
            this.tradeIdChanged = true;
        }
    }
    setAccountId(value) {
        this.accountIdChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setAccountId(value);
            this.accountIdChanged = true;
        }
    }
    setAccountName(value) {
        this.accountNameChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setAccountName(value);
            this.accountNameChanged = true;
        }
    }
    setAccountKind(value) {
        this.accountKindChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setAccountKind(value);
            this.accountKindChanged = true;
        }
    }
    setOfferId(value) {
        this.offerIdChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setOfferId(value);
            this.offerIdChanged = true;
        }
    }
    setAmount(value) {
        this.amountChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setAmount(variantCast.castToInt(variantCast.fromString(value)));
            this.amountChanged = true;
        }
    }
    setBuySell(value) {
        this.buySellChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setBuySell(value);
            this.buySellChanged = true;
        }
    }
    setOpenRate(value) {
        this.openRateChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setOpenRate(variantCast.castToReal(variantCast.fromString(value)));
            this.openRateChanged = true;
        }
    }
    setOpenTime(value) {
        this.openTimeChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setOpenTime(DateTimeParser.parseDateTimeFromString(value));
            this.openTimeChanged = true;
        }
    }
    setOpenQuoteId(value) {
        this.openQuoteIdChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setOpenQuoteId(value);
            this.openQuoteIdChanged = true;
        }
    }
    setOpenOrderId(value) {
        this.openOrderIdChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setOpenOrderId(value);
            this.openOrderIdChanged = true;
        }
    }
    setDividends(value) {
        this.dividendsChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            if (stdlib.len(value) == 0) {
                this.openPositionBuilder.setDividends(0.0);
                return;
            }
            this.openPositionBuilder.setDividends(variantCast.castToReal(variantCast.fromString(value)));
            this.dividendsChanged = true;
        }
    }
    setOpenOrderReqId(value) {
        this.openOrderReqIdChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setOpenOrderReqId(value);
            this.openOrderReqIdChanged = true;
        }
    }
    setOpenOrderRequestTXT(value) {
        this.openOrderRequestTXTChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setOpenOrderRequestTXT(value);
            this.openOrderRequestTXTChanged = true;
        }
    }
    setCommission(value) {
        this.commissionChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setCommission(variantCast.castToReal(variantCast.fromString(value)));
            this.commissionChanged = true;
        }
    }
    setRolloverInterest(value) {
        this.rolloverInterestChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setRolloverInterest(variantCast.castToReal(variantCast.fromString(value)));
            this.rolloverInterestChanged = true;
        }
    }
    setTradeIdOrigin(value) {
        this.tradeIdOriginChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setTradeIdOrigin(value);
            this.tradeIdOriginChanged = true;
        }
    }
    setValueDate(value) {
        this.valueDateChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setValueDate(value);
            this.valueDateChanged = true;
        }
    }
    setParties(value) {
        this.partiesChanged = false;
        if (FieldChangeChecker.isFieldChange(value)) {
            this.openPositionBuilder.setParties(value);
            this.partiesChanged = true;
        }
    }
    build() {
        this.openPosition = this.openPositionBuilder.build();
        return this;
    }
}
module.exports = OpenPositionUpdateBuilder
