// @gehtsoft/traduAPI-node/lib/OrderBuilder.js
'use strict';
let Order = require('./Order.js');
//@@DocBrief("The builder of the Order class")
class OrderBuilder extends Order {
    setOrderId(orderId) {
        this.orderId = orderId;
    }
    setRequestId(requestId) {
        this.requestId = requestId;
    }
    setRate(rate) {
        this.rate = rate;
    }
    setExecutionRate(executionRate) {
        this.executionRate = executionRate;
    }
    setRateMin(rateMin) {
        this.rateMin = rateMin;
    }
    setRateMax(rateMax) {
        this.rateMax = rateMax;
    }
    setTradeId(tradeId) {
        this.tradeId = tradeId;
    }
    setAccountId(accountId) {
        this.accountId = accountId;
    }
    setAccountName(accountName) {
        this.accountName = accountName;
    }
    setOfferId(offerId) {
        this.offerId = offerId;
    }
    setNetQuantity(netQuantity) {
        this.netQuantity = netQuantity;
    }
    setBuySell(buySell) {
        this.buySell = buySell;
    }
    setStage(stage) {
        this.stage = stage;
    }
    setType(type) {
        this.type = type;
    }
    setStatus(status) {
        this.status = status;
    }
    setStatusTime(statusTime) {
        this.statusTime = statusTime;
    }
    setAmount(amount) {
        this.amount = amount;
    }
    setLifetime(lifetime) {
        this.lifetime = lifetime;
    }
    setAtMarket(atMarket) {
        this.atMarket = atMarket;
    }
    setTrailStep(trailStep) {
        this.trailStep = trailStep;
    }
    setTrailRate(trailRate) {
        this.trailRate = trailRate;
    }
    setTimeInForce(timeInForce) {
        this.timeInForce = timeInForce;
    }
    setAccountKind(accountKind) {
        this.accountKind = accountKind;
    }
    setRequestTxt(requestTxt) {
        this.requestTxt = requestTxt;
    }
    setContingentOrderID(contingentOrderID) {
        this.contingentOrderID = contingentOrderID;
    }
    setContingencyType(contingencyType) {
        this.contingencyType = contingencyType;
    }
    setPrimaryId(primaryId) {
        this.primaryId = primaryId;
    }
    setOriginAmount(originAmount) {
        this.originAmount = originAmount;
    }
    setFilledAmount(filledAmount) {
        this.filledAmount = filledAmount;
    }
    setWorkingIndicator(workingIndicator) {
        this.workingIndicator = workingIndicator;
    }
    setPegType(pegType) {
        this.pegType = pegType;
    }
    setPegOffset(pegOffset) {
        this.pegOffset = pegOffset;
    }
    setPegOffsetMin(pegOffsetMin) {
        this.pegOffsetMin = pegOffsetMin;
    }
    setPegOffsetMax(pegOffsetMax) {
        this.pegOffsetMax = pegOffsetMax;
    }
    setExpireDate(expireDate) {
        this.expireDate = expireDate;
    }
    setValueDate(valueDate) {
        this.valueDate = valueDate;
    }
    setParties(parties) {
        this.parties = parties;
    }
    setSide(side) {
        this.side = side;
    }
    setStop(stop) {
        this.stop = stop;
    }
    setLimit(limit) {
        this.limit = limit;
    }
    setStopOrderId(stopOrderId) {
        this.stopOrderId = stopOrderId;
    }
    setLimitOrderId(limitOrderId) {
        this.limitOrderId = limitOrderId;
    }
    setTypeStop(typeStop) {
        this.typeStop = typeStop;
    }
    setTypeLimit(typeLimit) {
        this.typeLimit = typeLimit;
    }
    setStopTrailStep(stopTrailStep) {
        this.stopTrailStep = stopTrailStep;
    }
    setStopTrailRate(stopTrailRate) {
        this.stopTrailRate = stopTrailRate;
    }
    setStopMove(stopMove) {
        this.stopMove = stopMove;
    }
    build() {
        return this;
    }
}
module.exports = OrderBuilder
