// @gehtsoft/traduAPI-node/lib/OrderCalculatedFields.js
'use strict';
class OrderCalculatedFields {
    stop;
    limit;
    stopOrderId;
    limitOrderId;
    typeStop;
    typeLimit;
    stopTrailStep;
    stopTrailRate;
    orderTrailingStopData;
    constructor() {
        this.stop = 0.0;
        this.limit = 0.0;
        this.stopOrderId = null;
        this.limitOrderId = null;
        this.typeStop = 0;
        this.typeLimit = 0;
        this.stopTrailStep = 0;
        this.stopTrailRate = 0.0;
        this.orderTrailingStopData = null;
    }
    getStop() {
        return this.stop;
    }
    getLimit() {
        return this.limit;
    }
    getStopOrderId() {
        return this.stopOrderId;
    }
    getLimitOrderId() {
        return this.limitOrderId;
    }
    getTypeStop() {
        return this.typeStop;
    }
    getTypeLimit() {
        return this.typeLimit;
    }
    getStopTrailStep() {
        return this.stopTrailStep;
    }
    getStopTrailRate() {
        return this.stopTrailRate;
    }
    getOrderTrailingStopData() {
        return this.orderTrailingStopData;
    }
}
module.exports = OrderCalculatedFields
