// @gehtsoft/traduAPI-node/lib/OrderCalculatedFieldsBuilder.js
'use strict';
let OrderCalculatedFields = require('./OrderCalculatedFields.js');
class OrderCalculatedFieldsBuilder extends OrderCalculatedFields {
    setCalculatedStop(stop) {
        this.stop = stop;
    }
    setCalculatedLimit(limit) {
        this.limit = limit;
    }
    setCalculatedStopOrderId(stopOrderId) {
        this.stopOrderId = stopOrderId;
    }
    setCalculatedLimitOrderId(limitOrderId) {
        this.limitOrderId = limitOrderId;
    }
    setCalculatedTypeStop(typeStop) {
        this.typeStop = typeStop;
    }
    setCalculatedTypeLimit(typeLimit) {
        this.typeLimit = typeLimit;
    }
    setCalculatedStopTrailStep(stopTrailStep) {
        this.stopTrailStep = stopTrailStep;
    }
    setCalculatedStopTrailRate(stopTrailRate) {
        this.stopTrailRate = stopTrailRate;
    }
    setOrderTrailingStopData(orderTrailingStopData) {
        this.orderTrailingStopData = orderTrailingStopData;
    }
    build() {
        return this;
    }
}
module.exports = OrderCalculatedFieldsBuilder
