// @gehtsoft/traduAPI-node/lib/OrderUpdate.js
'use strict';
class OrderUpdate {
    order;
    orderIdChanged;
    requestIdChanged;
    rateChanged;
    executionRateChanged;
    rateMinChanged;
    rateMaxChanged;
    tradeIdChanged;
    accountIdChanged;
    accountNameChanged;
    offerIdChanged;
    netQuantityChanged;
    buySellChanged;
    stageChanged;
    typeChanged;
    statusChanged;
    statusTimeChanged;
    amountChanged;
    lifetimeChanged;
    atMarketChanged;
    trailStepChanged;
    trailRateChanged;
    timeInForceChanged;
    accountKindChanged;
    requestTxtChanged;
    contingentOrderIDChanged;
    contingencyTypeChanged;
    primaryIdChanged;
    originAmountChanged;
    filledAmountChanged;
    workingIndicatorChanged;
    pegTypeChanged;
    pegOffsetChanged;
    pegOffsetMinChanged;
    pegOffsetMaxChanged;
    expireDateChanged;
    valueDateChanged;
    partiesChanged;
    sideChanged;
    stopChanged;
    limitChanged;
    stopOrderIdChanged;
    limitOrderIdChanged;
    typeStopChanged;
    typeLimitChanged;
    stopTrailStepChanged;
    stopTrailRateChanged;
    getOrder() {
        return this.order;
    }
    isOrderIdChanged() {
        return this.orderIdChanged;
    }
    isRequestIdChanged() {
        return this.requestIdChanged;
    }
    isRateChanged() {
        return this.rateChanged;
    }
    isExecutionRateChanged() {
        return this.executionRateChanged;
    }
    isRateMinChanged() {
        return this.rateMinChanged;
    }
    isRateMaxChanged() {
        return this.rateMaxChanged;
    }
    isTradeIdChanged() {
        return this.tradeIdChanged;
    }
    isAccountIdChanged() {
        return this.accountIdChanged;
    }
    isAccountNameChanged() {
        return this.accountNameChanged;
    }
    isOfferIdChanged() {
        return this.offerIdChanged;
    }
    isNetQuantityChanged() {
        return this.netQuantityChanged;
    }
    isBuySellChanged() {
        return this.buySellChanged;
    }
    isStageChanged() {
        return this.stageChanged;
    }
    isTypeChanged() {
        return this.typeChanged;
    }
    isStatusChanged() {
        return this.statusChanged;
    }
    isStatusTimeChanged() {
        return this.statusTimeChanged;
    }
    isAmountChanged() {
        return this.amountChanged;
    }
    isLifetimeChanged() {
        return this.lifetimeChanged;
    }
    isAtMarketChanged() {
        return this.atMarketChanged;
    }
    isTrailStepChanged() {
        return this.trailStepChanged;
    }
    isTrailRateChanged() {
        return this.trailRateChanged;
    }
    isTimeInForceChanged() {
        return this.timeInForceChanged;
    }
    isAccountKindChanged() {
        return this.accountKindChanged;
    }
    isRequestTxtChanged() {
        return this.requestTxtChanged;
    }
    isContingentOrderIDChanged() {
        return this.contingentOrderIDChanged;
    }
    isContingencyTypeChanged() {
        return this.contingencyTypeChanged;
    }
    isPrimaryIdChanged() {
        return this.primaryIdChanged;
    }
    isOriginAmountChanged() {
        return this.originAmountChanged;
    }
    isFilledAmountChanged() {
        return this.filledAmountChanged;
    }
    isWorkingIndicatorChanged() {
        return this.workingIndicatorChanged;
    }
    isPegTypeChanged() {
        return this.pegTypeChanged;
    }
    isPegOffsetChanged() {
        return this.pegOffsetChanged;
    }
    isPegOffsetMinChanged() {
        return this.pegOffsetMinChanged;
    }
    isPegOffsetMaxChanged() {
        return this.pegOffsetMaxChanged;
    }
    isExpireDateChanged() {
        return this.expireDateChanged;
    }
    isValueDateChanged() {
        return this.valueDateChanged;
    }
    isPartiesChanged() {
        return this.partiesChanged;
    }
    isSideChanged() {
        return this.sideChanged;
    }
    isStopChanged() {
        return this.stopChanged;
    }
    isLimitChanged() {
        return this.limitChanged;
    }
    isStopOrderIdChanged() {
        return this.stopOrderIdChanged;
    }
    isLimitOrderIdChanged() {
        return this.limitOrderIdChanged;
    }
    isTypeStopChanged() {
        return this.typeStopChanged;
    }
    isTypeLimitChanged() {
        return this.typeLimitChanged;
    }
    isStopTrailStepChanged() {
        return this.stopTrailStepChanged;
    }
    isStopTrailRateChanged() {
        return this.stopTrailRateChanged;
    }
}
module.exports = OrderUpdate
