// @gehtsoft/traduAPI-node/lib/OrderRequestBuilderFactory.js
'use strict';
let MarketOrderRequestBuilder = require('./MarketOrderRequestBuilder.js');
let LimitOrderRequestBuilder = require('./LimitOrderRequestBuilder.js');
let StopOrderRequestBuilder = require('./StopOrderRequestBuilder.js');
let CloseMarketOrderRequestBuilder = require('./CloseMarketOrderRequestBuilder.js');
let EntryOrderRequestBuilder = require('./EntryOrderRequestBuilder.js');
let ChangeOrderRequestBuilder = require('./ChangeOrderRequestBuilder.js');
let CloseAllPositionsRequestBuilder = require('./CloseAllPositionsRequestBuilder.js');
//@@DocBrief("The factory for creating order request builders")
class OrderRequestBuilderFactory {
    createMarketOrderRequestBuilder() {
        return new MarketOrderRequestBuilder();
    }
    createLimitOrderRequestBuilder() {
        return new LimitOrderRequestBuilder();
    }
    createStopOrderRequestBuilder() {
        return new StopOrderRequestBuilder();
    }
    createCloseMarketOrderRequestBuilder() {
        return new CloseMarketOrderRequestBuilder();
    }
    createEntryOrderRequestBuilder() {
        return new EntryOrderRequestBuilder();
    }
    createChangeOrderRequestBuilder() {
        return new ChangeOrderRequestBuilder();
    }
    createCloseAllPositionsRequestBuilder() {
        return new CloseAllPositionsRequestBuilder();
    }
}
module.exports = OrderRequestBuilderFactory
