// @gehtsoft/traduAPI-node/lib/Order.js
'use strict';
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
//@@DocInclude
//@@DocBrief("The class provides access to order information")
class Order {
    orderId;
    requestId;
    rate;
    executionRate;
    rateMin;
    rateMax;
    tradeId;
    accountId;
    accountName;
    offerId;
    netQuantity;
    buySell;
    stage;
    type;
    status;
    statusTime;
    amount;
    lifetime;
    atMarket;
    trailStep;
    trailRate;
    timeInForce;
    accountKind;
    requestTxt;
    contingentOrderID;
    contingencyType;
    primaryId;
    originAmount;
    filledAmount;
    workingIndicator;
    pegType;
    pegOffset;
    pegOffsetMin;
    pegOffsetMax;
    expireDate;
    valueDate;
    parties;
    side;
    stop;
    limit;
    stopOrderId;
    limitOrderId;
    typeStop;
    typeLimit;
    stopTrailStep;
    stopTrailRate;
    stopMove;
    constructor() {
        this.orderId = null;
        this.requestId = null;
        this.rate = 0.0;
        this.executionRate = 0.0;
        this.rateMin = 0.0;
        this.rateMax = 0.0;
        this.tradeId = null;
        this.accountId = null;
        this.accountName = null;
        this.offerId = null;
        this.netQuantity = false;
        this.buySell = null;
        this.stage = null;
        this.type = null;
        this.status = null;
        this.statusTime = stdlib.mkdate(1900, 1, 1);
        this.amount = 0;
        this.lifetime = 0.0;
        this.atMarket = 0.0;
        this.trailStep = 0;
        this.trailRate = 0.0;
        this.timeInForce = null;
        this.accountKind = null;
        this.requestTxt = null;
        this.contingentOrderID = null;
        this.contingencyType = 0;
        this.primaryId = null;
        this.originAmount = 0;
        this.filledAmount = 0;
        this.workingIndicator = false;
        this.pegType = null;
        this.pegOffset = 0.0;
        this.pegOffsetMin = 0.0;
        this.pegOffsetMax = 0.0;
        this.expireDate = stdlib.mkdate(1900, 1, 1);
        this.valueDate = null;
        this.parties = null;
        this.side = 0;
        this.stop = 0.0;
        this.limit = 0.0;
        this.stopOrderId = null;
        this.limitOrderId = null;
        this.typeStop = 0;
        this.typeLimit = 0;
        this.stopTrailStep = 0;
        this.stopTrailRate = 0.0;
        this.stopMove = 0.0;
    }
    getOrderId() {
        return this.orderId;
    }
    getRequestId() {
        return this.requestId;
    }
    getRate() {
        return this.rate;
    }
    getExecutionRate() {
        return this.executionRate;
    }
    getRateMin() {
        return this.rateMin;
    }
    getRateMax() {
        return this.rateMax;
    }
    getTradeId() {
        return this.tradeId;
    }
    getAccountId() {
        return this.accountId;
    }
    getAccountName() {
        return this.accountName;
    }
    getOfferId() {
        return this.offerId;
    }
    getNetQuantity() {
        return this.netQuantity;
    }
    getBuySell() {
        return this.buySell;
    }
    getStage() {
        return this.stage;
    }
    getType() {
        return this.type;
    }
    getStatus() {
        return this.status;
    }
    getStatusTime() {
        return this.statusTime;
    }
    getAmount() {
        return this.amount;
    }
    getLifetime() {
        return this.lifetime;
    }
    getAtMarket() {
        return this.atMarket;
    }
    getTrailStep() {
        return this.trailStep;
    }
    getTrailRate() {
        return this.trailRate;
    }
    getTimeInForce() {
        return this.timeInForce;
    }
    getAccountKind() {
        return this.accountKind;
    }
    getRequestTxt() {
        return this.requestTxt;
    }
    getContingentOrderID() {
        return this.contingentOrderID;
    }
    getContingencyType() {
        return this.contingencyType;
    }
    getPrimaryId() {
        return this.primaryId;
    }
    getOriginAmount() {
        return this.originAmount;
    }
    getFilledAmount() {
        return this.filledAmount;
    }
    getWorkingIndicator() {
        return this.workingIndicator;
    }
    getPegType() {
        return this.pegType;
    }
    getPegOffset() {
        return this.pegOffset;
    }
    getPegOffsetMin() {
        return this.pegOffsetMin;
    }
    getPegOffsetMax() {
        return this.pegOffsetMax;
    }
    getExpireDate() {
        return this.expireDate;
    }
    getValueDate() {
        return this.valueDate;
    }
    getParties() {
        return this.parties;
    }
    getSide() {
        return this.side;
    }
    getStop() {
        return this.stop;
    }
    getLimit() {
        return this.limit;
    }
    getStopOrderId() {
        return this.stopOrderId;
    }
    getLimitOrderId() {
        return this.limitOrderId;
    }
    getTypeStop() {
        return this.typeStop;
    }
    getTypeLimit() {
        return this.typeLimit;
    }
    getStopTrailStep() {
        return this.stopTrailStep;
    }
    getStopTrailRate() {
        return this.stopTrailRate;
    }
    getStopMove() {
        return this.stopMove;
    }
}
module.exports = Order
