// @gehtsoft/traduAPI-node/lib/PdasMessageFieldFactory.js
'use strict';
let PdasMessageGroup = require('./PdasMessageGroup.js');
let PdasMessageFieldBuilder = require('./PdasMessageFieldBuilder.js');
let PdasMessageListBuilder = require('./PdasMessageListBuilder.js');
let PdasMessageBuilder = require('./PdasMessageBuilder.js');
let PdasMessageSerializer = require('./PdasMessageSerializer.js');
class PdasMessageFieldFactory {
    createGroup() {
        return new PdasMessageGroup();
    }
    createStringField(fieldTag, value) {
        let builder;
        builder = new PdasMessageFieldBuilder();
        return builder.buildString(fieldTag, value);
    }
    createDoubleField(fieldTag, value) {
        let builder;
        builder = new PdasMessageFieldBuilder();
        return builder.buildDouble(fieldTag, value);
    }
    createIntField(fieldTag, value) {
        let builder;
        builder = new PdasMessageFieldBuilder();
        return builder.buildInt(fieldTag, value);
    }
    createLongField(fieldTag, value) {
        let builder;
        builder = new PdasMessageFieldBuilder();
        return builder.buildLong(fieldTag, value);
    }
    createList(fieldTag) {
        let builder;
        builder = new PdasMessageListBuilder();
        return builder.build(fieldTag);
    }
    createMessage(msgType) {
        let builder;
        builder = new PdasMessageBuilder();
        return builder.build(msgType);
    }
    getMessageSerializer() {
        return new PdasMessageSerializer();
    }
}
module.exports = PdasMessageFieldFactory
