// @gehtsoft/traduAPI-node/lib/PdasGetAllOffersRequestMessage.js
'use strict';
let APdasGetDataRequestMessage = require('./APdasGetDataRequestMessage.js');
let PdasRequestCommand = require('./PdasRequestCommand.js');
let TransportMessageType = require('./TransportMessageType.js');
let MessageBuilder = require('./MessageBuilder.js');
class PdasGetAllOffersRequestMessage extends APdasGetDataRequestMessage {
    offerIds;
    requestNumber;
    localBuild(offerIds, tradingSession, requestNumber, transactTime) {
        this.offerIds = offerIds;
        this.requestNumber = requestNumber;
        return super.build(tradingSession, this.requestNumberGenerator, transactTime);
    }
    getCommand() {
        return PdasRequestCommand.GET_OFFERS;
    }
    getTransportMessageType() {
        return TransportMessageType.GetAllOffers;
    }
    getRequestNumber() {
        return this.requestNumber;
    }
    getAdditionalGroups(factory) {
        return [MessageBuilder.getOfferIdsGroup(factory, this.offerIds)];
    }
}
module.exports = PdasGetAllOffersRequestMessage
