// @gehtsoft/traduAPI-node/lib/PdasMessageGroup.js
'use strict';
let PdasMessageItemWithChild = require('./PdasMessageItemWithChild.js');
let PdasMessageFieldTag = require('./PdasMessageFieldTag.js');
class PdasMessageGroup {
    children;
    constructor() {
        this.children = new PdasMessageItemWithChild();
    }
    addChild(item) {
        this.children.addChild(item);
    }
    getName() {
        return this.children.getFieldValue(PdasMessageFieldTag.FXCM_PARAM_NAME);
    }
    getFieldValue(name) {
        return this.children.getFieldValue(name);
    }
    getList(name) {
        return this.children.getList(name);
    }
    getGroupParamValueByParamName(name) {
        return this.children.getGroupParamValueByParamName(name);
    }
    getFields() {
        return this.children.getFields();
    }
    getGroups() {
        return this.children.getGroups();
    }
    length() {
        return this.children.length();
    }
    serialize() {
        let result;
        result = "<g s=\"" + (this.children.length()).toString() + "\">" + this.children.serialize() + "</g>";
        return result;
    }
    getElementType() {
        return "g";
    }
}
module.exports = PdasMessageGroup
