// @gehtsoft/traduAPI-node/lib/PositionsSummaryBuilder.js
'use strict';
let PositionsSummary = require('./PositionsSummary.js');
//@@DocBrief("The builder of the PositionsSummary class")
class PositionsSummaryBuilder extends PositionsSummary {
    setInstrument(instrument) {
        this.instrument = instrument;
    }
    setOfferId(offerId) {
        this.offerId = offerId;
    }
    setSellNetPL(sellNetPL) {
        this.sellNetPL = sellNetPL;
    }
    setSellPipPL(sellPipPL) {
        this.sellPipPL = sellPipPL;
    }
    setSellUsedMr(sellUsedMr) {
        this.sellUsedMr = sellUsedMr;
    }
    setSellAmount(sellAmount) {
        this.sellAmount = sellAmount;
    }
    setSellAvgOpen(sellAvgOpen) {
        this.sellAvgOpen = sellAvgOpen;
    }
    setSellClose(sellClose) {
        this.sellClose = sellClose;
    }
    setBuyNetPL(buyNetPL) {
        this.buyNetPL = buyNetPL;
    }
    setBuyPipPL(buyPipPL) {
        this.buyPipPL = buyPipPL;
    }
    setBuyUsedMr(buyUsedMr) {
        this.buyUsedMr = buyUsedMr;
    }
    setBuyAmount(buyAmount) {
        this.buyAmount = buyAmount;
    }
    setBuyAvgOpen(buyAvgOpen) {
        this.buyAvgOpen = buyAvgOpen;
    }
    setBuyClose(buyClose) {
        this.buyClose = buyClose;
    }
    setAmount(amount) {
        this.amount = amount;
    }
    setGrossPL(grossPL) {
        this.grossPL = grossPL;
    }
    setBuyGrossPL(buyGrossPL) {
        this.buyGrossPL = buyGrossPL;
    }
    setSellGrossPL(sellGrossPL) {
        this.sellGrossPL = sellGrossPL;
    }
    setRolloverInterest(rolloverInterest) {
        this.rolloverInterest = rolloverInterest;
    }
    setNetPL(netPL) {
        this.netPL = netPL;
    }
    setNumberOfPositions(numberOfPositions) {
        this.numberOfPositions = numberOfPositions;
    }
    setSortOrder(sortOrder) {
        this.sortOrder = sortOrder;
    }
    setDividends(dividends) {
        this.dividends = dividends;
    }
    build() {
        return this;
    }
}
module.exports = PositionsSummaryBuilder
