// @gehtsoft/traduAPI-node/lib/PositionsSummary.js
'use strict';
//@@DocInclude
//@@DocBrief("The class provides access to summary information about all open positions in one instrument")
class PositionsSummary {
    instrument;
    offerId;
    sellNetPL;
    sellPipPL;
    sellUsedMr;
    sellAmount;
    sellAvgOpen;
    sellClose;
    buyNetPL;
    buyPipPL;
    buyUsedMr;
    buyAmount;
    buyAvgOpen;
    buyClose;
    amount;
    grossPL;
    buyGrossPL;
    sellGrossPL;
    rolloverInterest;
    netPL;
    numberOfPositions;
    sortOrder;
    dividends;
    constructor() {
        this.instrument = "";
        this.offerId = "";
        this.sellNetPL = 0.0;
        this.sellPipPL = 0.0;
        this.sellUsedMr = 0.0;
        this.sellAmount = 0;
        this.sellAvgOpen = 0.0;
        this.sellClose = 0.0;
        this.buyNetPL = 0.0;
        this.buyPipPL = 0.0;
        this.buyUsedMr = 0.0;
        this.buyAmount = 0;
        this.buyAvgOpen = 0.0;
        this.buyClose = 0.0;
        this.amount = 0;
        this.grossPL = 0.0;
        this.buyGrossPL = 0.0;
        this.sellGrossPL = 0.0;
        this.rolloverInterest = 0.0;
        this.netPL = 0.0;
        this.numberOfPositions = 0;
        this.sortOrder = 0;
        this.dividends = 0.0;
    }
    getInstrument() {
        return this.instrument;
    }
    getOfferId() {
        return this.offerId;
    }
    getSellNetPL() {
        return this.sellNetPL;
    }
    getSellPipPL() {
        return this.sellPipPL;
    }
    getSellUsedMr() {
        return this.sellUsedMr;
    }
    getSellAmount() {
        return this.sellAmount;
    }
    getSellAvgOpen() {
        return this.sellAvgOpen;
    }
    getSellClose() {
        return this.sellClose;
    }
    getBuyNetPL() {
        return this.buyNetPL;
    }
    getBuyPipPL() {
        return this.buyPipPL;
    }
    getBuyUsedMr() {
        return this.buyUsedMr;
    }
    getBuyAmount() {
        return this.buyAmount;
    }
    getBuyAvgOpen() {
        return this.buyAvgOpen;
    }
    getBuyClose() {
        return this.buyClose;
    }
    getAmount() {
        return this.amount;
    }
    getGrossPL() {
        return this.grossPL;
    }
    getBuyGrossPL() {
        return this.buyGrossPL;
    }
    getSellGrossPL() {
        return this.sellGrossPL;
    }
    getRolloverInterest() {
        return this.rolloverInterest;
    }
    getNetPL() {
        return this.netPL;
    }
    getNumberOfPositions() {
        return this.numberOfPositions;
    }
    getSortOrder() {
        return this.sortOrder;
    }
    getDividends() {
        return this.dividends;
    }
}
module.exports = PositionsSummary
