// @gehtsoft/traduAPI-node/lib/PriceFeedCommandBuilder.js
'use strict';
let PriceFeedCommand = require('./PriceFeedCommand.js');
class PriceFeedCommandBuilder extends PriceFeedCommand {
    setSessionStorage(sessionStorage) {
        this.sessionStorage = sessionStorage;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setRequestNumber(requestNumber) {
        this.requestNumber = requestNumber;
    }
    setInstruments(instruments) {
        this.instruments = instruments;
    }
    setCommandDescriptor(descriptor) {
        this.descriptor = descriptor;
    }
    setDxFeedNamesProvider(dxFeedNamesProvider) {
        this.dxFeedNamesProvider = dxFeedNamesProvider;
    }
    setPriceFeedCommandCallback(priceFeedCommandCallback) {
        this.priceFeedCommandCallback = priceFeedCommandCallback;
    }
    build() {
        return this;
    }
}
module.exports = PriceFeedCommandBuilder
