// @gehtsoft/traduAPI-node/lib/PriceFeedCommandDescriptor.js
'use strict';
let list = require('@gehtsoft/LuaxStdlib-node').list;
class PriceFeedCommandDescriptor {
    instruments;
    subscriptionRequestType;
    marketDepth;
    mdUpdateType;
    noMdUpdateTypes;
    fxcmTimingInterval;
    fxcmDasMessagePriceStream;
    fxcmTimingIntervalEx;
    fxcmIncludeWeekends;
    constructor() {
        this.subscriptionRequestType = "";
        this.marketDepth = 0;
        this.mdUpdateType = 0;
        this.fxcmTimingInterval = "";
        this.fxcmDasMessagePriceStream = "";
        this.fxcmTimingIntervalEx = "";
        this.fxcmIncludeWeekends = false;
        this.noMdUpdateTypes = "";
        this.instruments = new list();
    }
    getInstrumentDescriptors() {
        let instrumentDescriptors, i;
        instrumentDescriptors = new Array(this.instruments.length());
        for (i = 0; i <= this.instruments.length() - 1; i += 1) {
            instrumentDescriptors[i] = this.instruments.get(i);
        }
        return instrumentDescriptors;
    }
    getSubscriptionRequestType() {
        return this.subscriptionRequestType;
    }
    getMarketDepth() {
        return this.marketDepth;
    }
    getMdUpdateType() {
        return this.mdUpdateType;
    }
    getNoMdUpdateTypes() {
        return this.noMdUpdateTypes;
    }
    getFxcmTimingInterval() {
        return this.fxcmTimingInterval;
    }
    getFxcmDasMessagePriceStream() {
        return this.fxcmDasMessagePriceStream;
    }
    getFxcmTimingIntervalEx() {
        return this.fxcmTimingIntervalEx;
    }
    getFxcmIncludeWeekends() {
        return this.fxcmIncludeWeekends;
    }
}
module.exports = PriceFeedCommandDescriptor
