// @gehtsoft/traduAPI-node/lib/PriceHistoryRequest.js
'use strict';
//@@DocBrief("The class provides information for a price history request")
class PriceHistoryRequest {
    bidAdjustment;
    askAdjustment;
    instrument;
    instrumentDescriptor;
    timeframe;
    from;
    to;
    quotesCount;
    constructor() {
        this.bidAdjustment = 0;
        this.askAdjustment = 0;
    }
    getInstrument() {
        return this.instrument;
    }
    getInstrumentDescriptor() {
        return this.instrumentDescriptor;
    }
    getTimeframe() {
        return this.timeframe;
    }
    getFrom() {
        return this.from;
    }
    getTo() {
        return this.to;
    }
    getQuotesCount() {
        return this.quotesCount;
    }
    getBidAdjustment() {
        return this.bidAdjustment;
    }
    setBidAdjustment(bidAdjustment) {
        this.bidAdjustment = bidAdjustment;
    }
    getAskAdjustment() {
        return this.askAdjustment;
    }
    setAskAdjustment(askAdjustment) {
        this.askAdjustment = askAdjustment;
    }
    static create(instrument, timeframe, from, to, quotesCount, instrumentDescriptor) {
        let result;
        result = new PriceHistoryRequest();
        result.instrument = instrument;
        result.timeframe = timeframe;
        result.from = from;
        result.to = to;
        result.quotesCount = quotesCount;
        result.instrumentDescriptor = instrumentDescriptor;
        return result;
    }
}
module.exports = PriceHistoryRequest
