// @gehtsoft/traduAPI-node/lib/PriceHistoryDxfRequest.js
'use strict';
//@@DocBrief("The class provides information for a price history request")
class PriceHistoryDxfRequest {
    instrument;
    timeframe;
    from;
    to;
    toFilled;
    quotesCount;
    getInstrument() {
        return this.instrument;
    }
    getTimeframe() {
        return this.timeframe;
    }
    getFrom() {
        return this.from;
    }
    getTo() {
        return this.to;
    }
    isToFilled() {
        return this.toFilled;
    }
    getQuotesCount() {
        return this.quotesCount;
    }
    static create(instrument, timeframe, from, to, quotesCount) {
        let result;
        result = new PriceHistoryDxfRequest();
        result.instrument = instrument;
        result.timeframe = timeframe;
        result.from = from;
        result.to = to;
        result.quotesCount = quotesCount;
        result.toFilled = true;
        return result;
    }
    static createWithoutTo(instrument, timeframe, from, quotesCount) {
        let result;
        result = new PriceHistoryDxfRequest();
        result.instrument = instrument;
        result.timeframe = timeframe;
        result.from = from;
        result.quotesCount = quotesCount;
        result.toFilled = false;
        return result;
    }
}
module.exports = PriceHistoryDxfRequest
