// @gehtsoft/traduAPI-node/lib/PriceFeedCommandDescriptorBuilder.js
'use strict';
let PriceFeedCommandDescriptor = require('./PriceFeedCommandDescriptor.js');
class PriceFeedCommandDescriptorBuilder extends PriceFeedCommandDescriptor {
    static get SNAPSHOT() { return "0"; }
    static get SUBSCRIBE() { return "1"; }
    static get UNSUBSCRIBE() { return "2"; }
    addInstrumentDescriptor(instrumentDescriptor) {
        this.instruments.add(instrumentDescriptor);
        return this;
    }
    setSubscriptionRequestType(subscriptionRequestType) {
        this.subscriptionRequestType = subscriptionRequestType;
        return this;
    }
    setMarketDepth(marketDepth) {
        this.marketDepth = marketDepth;
        return this;
    }
    setMdUpdateType(mdUpdateType) {
        this.mdUpdateType = mdUpdateType;
        return this;
    }
    setNoMdUpdateTypes(noMdUpdateTypes) {
        this.noMdUpdateTypes = noMdUpdateTypes;
        return this;
    }
    setFxcmTimingInterval(fxcmTimingInterval) {
        this.fxcmTimingInterval = fxcmTimingInterval;
        return this;
    }
    setFxcmDasMessagePriceStream(fxcmDasMessagePriceStream) {
        this.fxcmDasMessagePriceStream = fxcmDasMessagePriceStream;
        return this;
    }
    setFxcmTimingIntervalEx(fxcmTimingIntervalEx) {
        this.fxcmTimingIntervalEx = fxcmTimingIntervalEx;
        return this;
    }
    setIncludeWeekends(fxcmIncludeWeekends) {
        this.fxcmIncludeWeekends = fxcmIncludeWeekends;
        return this;
    }
    build() {
        return this;
    }
}
module.exports = PriceFeedCommandDescriptorBuilder
