// @gehtsoft/traduAPI-node/lib/RakeOutPricesCommandBuilder.js
'use strict';
let RakeOutPricesCommand = require('./RakeOutPricesCommand.js');
//@@DocBrief("The builder of the RakeOutPricesCommand class")
class RakeOutPricesCommandBuilder extends RakeOutPricesCommand {
    setPeriodInMilliseconds(periodInMilliseconds) {
        this.periodInMilliseconds = periodInMilliseconds;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setConnectionParameters(connectionParameters) {
        this.connectionParameters = connectionParameters;
    }
    setPriceTerminal(priceTerminal) {
        this.priceTerminal = priceTerminal;
    }
    build() {
        this.createSchedulers();
        return this;
    }
}
module.exports = RakeOutPricesCommandBuilder
