// @gehtsoft/traduAPI-node/lib/RakeOutPriceSessionCommandExecutor.js
'use strict';
class RakeOutPriceSessionCommandExecutor {
    static get DEFAULT_PERIOD_MILLISECOND() { return 300; }
    priceSessionCommandExecutor;
    pdasPriceTerminalControllerCallback;
    rakeOutController;
    connectionParameters;
    commandFactory;
    messageFactory;
    requestNumberGenerator;
    messageExecutor;
    messageRouter;
    priceSessionProvider;
    connectionStatusManager;
    priceTerminal;
    rakeOutControllerFactory;
    rakeOutPriceControllerCallback;
    constructor() {
        this.priceSessionCommandExecutor = null;
        this.pdasPriceTerminalControllerCallback = null;
        this.rakeOutController = null;
        this.connectionParameters = null;
        this.commandFactory = null;
        this.messageFactory = null;
        this.requestNumberGenerator = null;
        this.messageExecutor = null;
        this.messageRouter = null;
        this.priceSessionProvider = null;
        this.connectionStatusManager = null;
        this.priceTerminal = null;
    }
    setSuccessor(priceSessionCommandExecutor) {
        this.priceSessionCommandExecutor = priceSessionCommandExecutor;
    }
    execute(pdasPriceTerminalControllerCallback) {
        this.pdasPriceTerminalControllerCallback = pdasPriceTerminalControllerCallback;
        if (this.priceSessionProvider.getSession() != null) {
            this.rakeOutController = this.rakeOutControllerFactory.createRakeOutPrices(this.commandFactory, this.connectionParameters, this.priceSessionProvider, this.messageFactory, this.messageExecutor, this.messageRouter, this.connectionStatusManager, this.requestNumberGenerator, RakeOutPriceSessionCommandExecutor.DEFAULT_PERIOD_MILLISECOND, this.priceTerminal, this.rakeOutPriceControllerCallback);
            this.rakeOutController.start();
            if (this.priceSessionCommandExecutor != null) {
                this.priceSessionCommandExecutor.execute(pdasPriceTerminalControllerCallback);
            }
        }
        else {
            pdasPriceTerminalControllerCallback.onError("The pdas price session is empty.");
        }
    }
    static create(commandEnvironment, commandFactory, priceSessionProvider, connectionStatusManager, priceTerminal, rakeOutControllerFactory, rakeOutPriceControllerCallback) {
        let rakeOutPriceSessionCommandExecutor;
        rakeOutPriceSessionCommandExecutor = new RakeOutPriceSessionCommandExecutor();
        rakeOutPriceSessionCommandExecutor.connectionParameters = commandEnvironment.getConnectionParameters();
        rakeOutPriceSessionCommandExecutor.commandFactory = commandFactory;
        rakeOutPriceSessionCommandExecutor.messageFactory = commandEnvironment.getMessageFactory();
        rakeOutPriceSessionCommandExecutor.requestNumberGenerator = commandEnvironment.getRequestNumberGenerator();
        rakeOutPriceSessionCommandExecutor.messageExecutor = commandEnvironment.getMessageExecutor();
        rakeOutPriceSessionCommandExecutor.messageRouter = commandEnvironment.getMessageRouter();
        rakeOutPriceSessionCommandExecutor.priceSessionProvider = priceSessionProvider;
        rakeOutPriceSessionCommandExecutor.connectionStatusManager = connectionStatusManager;
        rakeOutPriceSessionCommandExecutor.priceTerminal = priceTerminal;
        rakeOutPriceSessionCommandExecutor.rakeOutControllerFactory = rakeOutControllerFactory;
        rakeOutPriceSessionCommandExecutor.rakeOutPriceControllerCallback = rakeOutPriceControllerCallback;
        return rakeOutPriceSessionCommandExecutor;
    }
    stop() {
        if (this.rakeOutController != null) {
            this.rakeOutController.shutdown();
        }
    }
}
module.exports = RakeOutPriceSessionCommandExecutor
