// @gehtsoft/traduAPI-node/lib/RawOpenPositionBuilder.js
'use strict';
let RawOpenPosition = require('./RawOpenPosition.js');
//@@DocBrief("The builder of the OpenPosition class")
class RawOpenPositionBuilder extends RawOpenPosition {
    setTradeID(value) {
        this.tradeID = value;
    }
    setAccountId(value) {
        this.accountId = value;
    }
    setAccountName(value) {
        this.accountName = value;
    }
    setAccountKind(value) {
        this.accountKind = value;
    }
    setOfferId(value) {
        this.offerId = value;
    }
    setAmount(value) {
        this.amount = value;
    }
    setBuySell(value) {
        this.buySell = value;
    }
    setOpenRate(value) {
        this.openRate = value;
    }
    setOpenTime(value) {
        this.openTime = value;
    }
    setOpenQuoteId(value) {
        this.openQuoteId = value;
    }
    setOpenOrderId(value) {
        this.openOrderId = value;
    }
    setOpenOrderReqId(value) {
        this.openOrderReqId = value;
    }
    setOpenOrderRequestTXT(value) {
        this.openOrderRequestTXT = value;
    }
    setCommission(value) {
        this.commission = value;
    }
    setRolloverInterest(value) {
        this.rolloverInterest = value;
    }
    setTradeIdOrigin(value) {
        this.tradeIdOrigin = value;
    }
    setValueDate(value) {
        this.valueDate = value;
    }
    setParties(value) {
        this.parties = value;
    }
    setDividends(value) {
        this.dividends = value;
    }
    build() {
        return this;
    }
}
module.exports = RawOpenPositionBuilder
