// @gehtsoft/traduAPI-node/lib/RawOpenPosition.js
'use strict';
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
//@@DocInclude
//@@DocBrief("The class provides access to information about an open position, its received from the trading server without calculated fields.")
class RawOpenPosition {
    tradeID;
    accountId;
    accountName;
    accountKind;
    offerId;
    amount;
    buySell;
    openRate;
    openTime;
    openQuoteId;
    openOrderId;
    openOrderReqId;
    openOrderRequestTXT;
    commission;
    rolloverInterest;
    tradeIdOrigin;
    valueDate;
    parties;
    dividends;
    constructor() {
        this.tradeID = null;
        this.accountId = null;
        this.accountName = null;
        this.accountKind = null;
        this.offerId = null;
        this.amount = 0;
        this.buySell = null;
        this.openRate = 0.0;
        this.openTime = stdlib.nowlocal();
        this.openQuoteId = null;
        this.openOrderId = null;
        this.openOrderReqId = null;
        this.openOrderRequestTXT = null;
        this.commission = 0.0;
        this.dividends = 0.0;
        this.rolloverInterest = 0.0;
        this.tradeIdOrigin = null;
        this.valueDate = null;
        this.parties = null;
    }
    getTradeID() {
        return this.tradeID;
    }
    getAccountId() {
        return this.accountId;
    }
    getAccountName() {
        return this.accountName;
    }
    getAccountKind() {
        return this.accountKind;
    }
    getOfferId() {
        return this.offerId;
    }
    getAmount() {
        return this.amount;
    }
    getBuySell() {
        return this.buySell;
    }
    getOpenRate() {
        return this.openRate;
    }
    getOpenTime() {
        return this.openTime;
    }
    getOpenQuoteId() {
        return this.openQuoteId;
    }
    getOpenOrderId() {
        return this.openOrderId;
    }
    getOpenOrderReqId() {
        return this.openOrderReqId;
    }
    getOpenOrderRequestTXT() {
        return this.openOrderRequestTXT;
    }
    getCommission() {
        return this.commission;
    }
    getRolloverInterest() {
        return this.rolloverInterest;
    }
    getTradeIdOrigin() {
        return this.tradeIdOrigin;
    }
    getValueDate() {
        return this.valueDate;
    }
    getParties() {
        return this.parties;
    }
    getDividends() {
        return this.dividends;
    }
}
module.exports = RawOpenPosition
