// @gehtsoft/traduAPI-node/lib/RawOffer.js
'use strict';
//@@DocInclude
//@@DocBrief("The class provides access to raw offer information of a trading instrument, its received from the trading server without calculated fields.")
class RawOffer {
    bid;
    ask;
    high;
    low;
    offerId;
    bidTradable;
    askTradable;
    digits;
    pointSize;
    tradingStatus;
    constructor() {
        this.tradingStatus = "";
        this.offerId = "";
        this.bidTradable = "";
        this.askTradable = "";
        this.bid = 0.0;
        this.ask = 0.0;
        this.high = 0.0;
        this.low = 0.0;
        this.digits = 0;
        this.pointSize = 0.0;
    }
    getBid() {
        return this.bid;
    }
    getAsk() {
        return this.ask;
    }
    getHigh() {
        return this.high;
    }
    getLow() {
        return this.low;
    }
    getOfferId() {
        return this.offerId;
    }
    getBidTradable() {
        return this.bidTradable;
    }
    getAskTradable() {
        return this.askTradable;
    }
    getDigits() {
        return this.digits;
    }
    getPointSize() {
        return this.pointSize;
    }
    getTradingStatus() {
        return this.tradingStatus;
    }
}
module.exports = RawOffer
