// @gehtsoft/traduAPI-node/lib/RawOfferFromStringBuilder.js
'use strict';
let RawOffer = require('./RawOffer.js');
let variantCast = require('@gehtsoft/LuaxStdlib-node').variantCast;
let stdlib = require('@gehtsoft/LuaxStdlib-node').stdlib;
//@@DocBrief("The builder of the raw offer class")
class RawOfferFromStringBuilder extends RawOffer {
    setOfferId(offerId) {
        this.offerId = offerId;
    }
    setBidTradable(bidTradable) {
        this.bidTradable = bidTradable;
    }
    setAskTradable(askTradable) {
        this.askTradable = askTradable;
    }
    setHigh(high) {
        this.high = variantCast.castToReal(variantCast.fromString(high));
    }
    setLow(low) {
        this.low = variantCast.castToReal(variantCast.fromString(low));
    }
    setBid(bid) {
        this.bid = variantCast.castToReal(variantCast.fromString(bid));
    }
    setAsk(ask) {
        this.ask = variantCast.castToReal(variantCast.fromString(ask));
    }
    setDigits(digits) {
        this.digits = variantCast.castToInt(variantCast.fromString(digits));
    }
    setPointSize(pointSize) {
        let precision, cnt, result;
        result = 1;
        cnt = 0;
        precision = variantCast.castToInt(variantCast.fromString(pointSize));
        while(cnt < precision) {
            result = result * 0.1;
            cnt = cnt + 1;
        }
        this.pointSize = stdlib.roundInl(result, precision);
    }
    setTradingStatus(tradingStatus) {
        this.tradingStatus = tradingStatus;
    }
    build() {
        return this;
    }
}
module.exports = RawOfferFromStringBuilder
