// @gehtsoft/traduAPI-node/lib/RolloverProfilesStorage.js
'use strict';
let string_map = require('@gehtsoft/LuaxStdlib-node').string_map;
let RolloverMarkupDescriptor = require('./RolloverMarkupDescriptor.js');
//@@DocBrief("class provides access to a collection of rollover markup levels information")
class RolloverProfilesStorage {
    storage;
    constructor() {
        this.storage = new string_map();
    }
    clear() {
        this.storage.clear();
    }
    addRolloverProfiles(rolloverProfiles) {
        let i;
        for (i = 0; i <= rolloverProfiles.length - 1; i += 1) {
            this.addRolloverProfile(rolloverProfiles[i]);
        }
    }
    addRolloverProfile(rolloverProfile) {
        let rolloverDescriptorStorage, rolloverDescriptor;
        rolloverDescriptor = RolloverMarkupDescriptor.create(rolloverProfile.getRolloverSellMarkup(), rolloverProfile.getRolloverBuyMarkup());
        if (!this.storage.contains(rolloverProfile.getARPId())) {
            rolloverDescriptorStorage = new string_map();
            this.storage.set(rolloverProfile.getARPId(), rolloverDescriptorStorage);
        }
        else {
            rolloverDescriptorStorage = this.storage.get(rolloverProfile.getARPId());
        }
        rolloverDescriptorStorage.set(rolloverProfile.getOfferID(), rolloverDescriptor);
    }
    findRolloverMarkupDescriptor(rolloverProfileId, offerId) {
        let rolloverDescriptorStorage;
        if (!this.storage.contains(rolloverProfileId)) {
            return null;
        }
        else {
            rolloverDescriptorStorage = this.storage.get(rolloverProfileId);
            if (!rolloverDescriptorStorage.contains(offerId)) {
                return null;
            }
            else {
                return rolloverDescriptorStorage.get(offerId);
            }
        }
    }
}
module.exports = RolloverProfilesStorage
