// @gehtsoft/traduAPI-node/lib/SendOcoOrderCommandBuilder.js
'use strict';
let SendOcoOrderCommand = require('./SendOcoOrderCommand.js');
class SendOcoOrderCommandBuilder extends SendOcoOrderCommand {
    setTimeController(timeController) {
        this.timeController = timeController;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setTradingSession(tradingSession) {
        this.tradingSession = tradingSession;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setEntryOrderRequests(entryOrderRequests) {
        this.entryOrderRequests = entryOrderRequests;
    }
    setSides(sides) {
        this.sides = sides;
    }
    setClientRates(clientRates) {
        this.clientRates = clientRates;
    }
    setPointSizes(pointSizes) {
        this.pointSizes = pointSizes;
    }
    build() {
        return this;
    }
}
module.exports = SendOcoOrderCommandBuilder
