// @gehtsoft/traduAPI-node/lib/SendEntryOrderCommandBuilder.js
'use strict';
let SendEntryOrderCommand = require('./SendEntryOrderCommand.js');
class SendEntryOrderCommandBuilder extends SendEntryOrderCommand {
    setTimeController(timeController) {
        this.timeController = timeController;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setTradingSession(tradingSession) {
        this.tradingSession = tradingSession;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setEntryOrderRequest(entryOrderRequest) {
        this.entryOrderRequest = entryOrderRequest;
    }
    setSide(side) {
        this.side = side;
    }
    setClientRate(clientRate) {
        this.clientRate = clientRate;
    }
    setPointSize(pointSize) {
        this.pointSize = pointSize;
    }
    build() {
        return this;
    }
}
module.exports = SendEntryOrderCommandBuilder
