// @gehtsoft/traduAPI-node/lib/SendTrueMarketOrderCommandBuilder.js
'use strict';
let SendTrueMarketOrderCommand = require('./SendTrueMarketOrderCommand.js');
class SendTrueMarketOrderCommandBuilder extends SendTrueMarketOrderCommand {
    setTimeController(timeController) {
        this.timeController = timeController;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setTradingSession(tradingSession) {
        this.tradingSession = tradingSession;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setMarketOrderRequest(marketOrderRequest) {
        this.marketOrderRequest = marketOrderRequest;
    }
    setClientRate(clientRate) {
        this.clientRate = clientRate;
    }
    setPointSize(pointSize) {
        this.pointSize = pointSize;
    }
    build() {
        return this;
    }
}
module.exports = SendTrueMarketOrderCommandBuilder
