// @gehtsoft/traduAPI-node/lib/SendCloseMarketOrderCommandBuilder.js
'use strict';
let SendCloseMarketOrderCommand = require('./SendCloseMarketOrderCommand.js');
class SendCloseMarketOrderCommandBuilder extends SendCloseMarketOrderCommand {
    SendTrueMarketOrderCommandBuilder() {
        return void(0);
    }
    setTimeController(timeController) {
        this.timeController = timeController;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setTradingSession(tradingSession) {
        this.tradingSession = tradingSession;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setCloseMarketOrderRequest(closeMarketOrderRequest) {
        this.closeMarketOrderRequest = closeMarketOrderRequest;
    }
    setOpenPosition(openPosition) {
        this.openPosition = openPosition;
    }
    setClientRate(clientRate) {
        this.clientRate = clientRate;
    }
    setPointSize(pointSize) {
        this.pointSize = pointSize;
    }
    build() {
        return this;
    }
}
module.exports = SendCloseMarketOrderCommandBuilder
