// @gehtsoft/traduAPI-node/lib/SessionStorage.js
'use strict';
let PriceTerminalsManager = require('./PriceTerminalsManager.js');
//@@DocBrief("The interface stores received trading session.")
class SessionStorage {
    tradeSession;
    priceSession;
    priceTerminalsManager;
    constructor() {
        this.tradeSession = null;
        this.priceSession = null;
        this.priceTerminalsManager = new PriceTerminalsManager();
    }
    getMdtConnectionManager() {
        return this.priceTerminalsManager;
    }
    setTradingSession(session) {
        this.tradeSession = session;
    }
    setPriceSession(priceSession) {
        this.priceSession = priceSession;
    }
    getTradingSession() {
        return this.tradeSession;
    }
    getPriceTerminalInitializerByPriceStreamId(streamId) {
        return this.priceTerminalsManager.getPriceTerminalInitializer(streamId);
    }
    addPriceTerminalInitializer(streams, priceTerminalInitializer, addToMdtMap) {
        this.priceTerminalsManager.addPriceTerminalInitializer(streams, priceTerminalInitializer, addToMdtMap);
    }
}
module.exports = SessionStorage
