// @gehtsoft/traduAPI-node/lib/TerminalUrl.js
'use strict';
let string_map = require('@gehtsoft/LuaxStdlib-node').string_map;
let StringValueObject = require('./StringValueObject.js');
//@@DocBrief("The terminal URL class")
class TerminalUrl {
    name;
    httpProtocol;
    terminalProtocol;
    url;
    isSecureField;
    params;
    constructor() {
        this.name = "";
        this.httpProtocol = "";
        this.terminalProtocol = "";
        this.url = "";
        this.isSecureField = false;
        this.params = new string_map();
    }
    getName() {
        return this.name;
    }
    getUrl() {
        return this.url;
    }
    getTerminalProtocol() {
        return this.terminalProtocol;
    }
    getProtocol() {
        return this.httpProtocol;
    }
    isSecure() {
        return this.isSecureField;
    }
    getParam(_param) {
        if (this.params.contains(_param)) {
            return this.params.get(_param).get();
        }
        return "";
    }
    setParam(_key, _param) {
        this.params.set(_key, StringValueObject.create(_param));
    }
    getParamKeys() {
        return this.params.keys();
    }
}
module.exports = TerminalUrl
