// @gehtsoft/traduAPI-node/lib/TestRequestMdtCommandBuilder.js
'use strict';
let TestRequestMdtCommand = require('./TestRequestMdtCommand.js');
//@@DocBrief("The builder of the TestRequestPriceCommand class")
class TestRequestMdtCommandBuilder extends TestRequestMdtCommand {
    setPeriodInMilliseconds(periodInMilliseconds) {
        this.periodInMilliseconds = periodInMilliseconds;
    }
    setMessageExecutor(messageExecutor) {
        this.messageExecutor = messageExecutor;
    }
    setMessageFactory(messageFactory) {
        this.messageFactory = messageFactory;
    }
    setRequestNumberGenerator(requestNumberGenerator) {
        this.requestNumberGenerator = requestNumberGenerator;
    }
    setMessageRouter(messageRouter) {
        this.messageRouter = messageRouter;
    }
    setConnectionParameters(connectionParameters) {
        this.connectionParameters = connectionParameters;
    }
    setPriceTerminal(priceTerminal) {
        this.priceTerminal = priceTerminal;
    }
    build() {
        this.createSchedulers();
        return this;
    }
}
module.exports = TestRequestMdtCommandBuilder
