// @gehtsoft/traduAPI-node/lib/TradingTerminalBuilder.js
'use strict';
let TradingTerminal = require('./TradingTerminal.js');
//@@DocBrief("The builder of the TradingTerminal class")
class TradingTerminalBuilder extends TradingTerminal {
    setType(type) {
        this.terminal.type = type;
    }
    setName(name) {
        this.terminal.name = name;
    }
    setStatus(status) {
        this.terminal.status = status;
    }
    setInactiveText(inactiveText) {
        this.terminal.inactiveText = inactiveText;
    }
    setUrls(urls) {
        this.terminal.urls = urls;
    }
    setId(id) {
        this.terminal.id = id;
    }
    setSubId(subId) {
        this.terminal.subId = subId;
    }
    setDescription(description) {
        this.description = description;
    }
    setSalt(salt) {
        this.salt = salt;
    }
    setPriceTerminal(priceTerminal) {
        this.priceTerminal = priceTerminal;
    }
    setPriceTerminals(priceTerminals) {
        this.priceTerminals = priceTerminals;
    }
    setChartTerminal(chartTerminal) {
        this.chartTerminal = chartTerminal;
    }
    build() {
        return this;
    }
}
module.exports = TradingTerminalBuilder
