// @gehtsoft/traduAPI-node/lib/TradingTerminal.js
'use strict';
let Terminal = require('./Terminal.js');
//@@DocBrief("The trading terminal class")
class TradingTerminal {
    description;
    salt;
    priceTerminal;
    priceTerminals;
    chartTerminal;
    terminal;
    constructor() {
        this.terminal = new Terminal();
        this.priceTerminals = null;
        this.priceTerminal = null;
    }
    getType() {
        return this.terminal.getType();
    }
    getName() {
        return this.terminal.getName();
    }
    getStatus() {
        return this.terminal.getStatus();
    }
    getInactiveText() {
        return this.terminal.getInactiveText();
    }
    getUrls() {
        return this.terminal.getUrls();
    }
    getId() {
        return this.terminal.getId();
    }
    getSubId() {
        return this.terminal.getSubId();
    }
    getDescription() {
        return this.description;
    }
    getSalt() {
        return this.salt;
    }
    getStreams() {
        return new Array(0);
    }
    getPriceTerminals() {
        let result, i;
        if (this.priceTerminal != null) {
            return [this.priceTerminal];
        } else if (this.priceTerminals != null) {
            result = new Array(this.priceTerminals.length());
            for (i = 0; i <= this.priceTerminals.length() - 1; i += 1) {
                result[i] = this.priceTerminals.get(i);
            }
            return result;
        }
        return new Array(0);
    }
    getPriceTerminalByPriceStream(streamId) {
        let terminal, streams, i, y;
        if (this.priceTerminals == null) {
            return null;
        }
        for (i = 0; i <= this.priceTerminals.length() - 1; i += 1) {
            terminal = this.priceTerminals.get(i);
            streams = terminal.getStreams();
            for (y = 0; y <= streams.length - 1; y += 1) {
                if (streams[y] == streamId) {
                    return terminal;
                }
            }
        }
        return null;
    }
    getChartTerminal() {
        return this.chartTerminal;
    }
}
module.exports = TradingTerminal
