// @gehtsoft/traduAPI-node/lib/TrailingType.js
'use strict';
//@@DocInclude
//@@DocBrief("The class for the trailing type")
class TrailingType {
    static get TrailingDenied() { return 0; }
    static get TrailingDynamic() { return 1; }
    static get TrailingFluctuate() { return 2; }
    static get TrailingDynamicAndFluctuate() { return 3; }
    static get TrailingStopDynamic() { return 4; }
    static get TrailingStopFluctuate() { return 5; }
    trailingType;
    constructor() {
        this.trailingType = TrailingType.TrailingDenied;
    }
    setTrailingType(trailingType) {
        this.trailingType = trailingType;
    }
    isTrailingDenied() {
        return this.trailingType == TrailingType.TrailingDenied;
    }
    isTrailingDynamic() {
        return this.trailingType == TrailingType.TrailingDynamic;
    }
    isTrailingFluctuate() {
        return this.trailingType == TrailingType.TrailingFluctuate;
    }
    isTrailingDynamicAndFluctuate() {
        return this.trailingType == TrailingType.TrailingDynamicAndFluctuate;
    }
    isTrailingStopDynamic() {
        return this.trailingType == TrailingType.TrailingStopDynamic;
    }
    isTrailingStopFluctuate() {
        return this.trailingType == TrailingType.TrailingStopFluctuate;
    }
}
module.exports = TrailingType
