// @gehtsoft/traduAPI-node/lib/UserKind.js
'use strict';
//@@DocInclude
//@@DocBrief("This class defines a set of constants and methods to represent different types of users")
class UserKind {
    static get TRADER_NAME() { return "Trader"; }
    static get CUSTOMER_NAME() { return "Customer"; }
    static get DEALER_NAME() { return "Dealer"; }
    static get ADMIN_NAME() { return "Admin"; }
    static get MERCHANT_NAME() { return "Merchant"; }
    static get UNDEFINED_NAME() { return "Undefined"; }
    kind;
    static create(kind) {
        let result;
        result = new UserKind();
        result.kind = kind;
        return result;
    }
    isTrader() {
        return this.kind == "20";
    }
    isCustomer() {
        return this.kind == "22";
    }
    isDealer() {
        return this.kind == "24";
    }
    isAdmin() {
        return this.kind == "26";
    }
    isMerchant() {
        return this.kind == "30";
    }
    getName() {
        if (this.kind == "20") {
            return UserKind.TRADER_NAME;
        }
        if (this.kind == "22") {
            return UserKind.CUSTOMER_NAME;
        }
        if (this.kind == "24") {
            return UserKind.DEALER_NAME;
        }
        if (this.kind == "26") {
            return UserKind.ADMIN_NAME;
        }
        if (this.kind == "30") {
            return UserKind.MERCHANT_NAME;
        }
        return UserKind.UNDEFINED_NAME;
    }
}
module.exports = UserKind
